---
title: "trackViewer_workshop2020"
bibliography: "`r system.file('vignettes', 'bibliography.bib', package = 'integratedchipseqanalysisworkshop')`"
csl: "`r system.file('vignettes', 'nature.csl', package = 'integratedchipseqanalysisworkshop')`"
vignette: >
  %\VignetteIndexEntry{integratedchipseqanalysisworkshop_trackViewer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output: 
  bookdown::html_document2:
    theme: simplex
    toc: true
    toc_float: true
    toc_depth: 4
    fig_caption: true
---

```{r checkplatform, include=FALSE}
library(knitr)
suppressPackageStartupMessages({
  library(trackViewer)
  library(pander)
  library(VariantAnnotation)
  library(TxDb.Hsapiens.UCSC.hg19.knownGene)
  library(org.Hs.eg.db)
  library(GenomicFeatures)
  library(org.Mm.eg.db)
})
opts_chunk$set(eval=TRUE, fig.width=6, fig.height=2, 
               warning = FALSE, message = FALSE)
```

## Introduction

It is a routine way to show the distribution of mutation for genetic variations by lollipop-style (or needle-style) plot in a genome browser, along with a variety of genomic annotations, such as gene level or exon level models, CpG island, and so on. To show the SNP status or methylation data, a special plot, called lollipop plot or needle plot, is used to show the distribution along annotations. Many of currently available tools offline or online provide lollipop-like plot to show the mutation data such as cBioPortal Tools::MutationMapper[@gao2013integrative], EMBL-EBI::Pfam[@finn2006pfam], and BioJS::muts-needle-plot[@michael_p_schroeder_2015_14561], BiQ Analyzer[@bock2005biq], and Methylation plotter[@mallona2014methylation]. The cBioPortal Tools::MutationMapper is a well-known and easy to use on-line genome browser that can output high quality of figures with mutations by input tab-delimited mutation data. In R/Bioconductor, there are options to use the flexibility of the Rgraphics system to display Methylation data such as _MethVisual_[@zackay2010methvisual], _REDseq_[@zhu4redseq], and _GenVisR_[@skidmore2016genvisr]. 

**Table 1: Tools availble for lollipop plot. **
```{r echo=FALSE}
software <- c("MutationMapper[@gao2013integrative]", "Pfam[@finn2006pfam]", "muts-needle-plot[@michael_p_schroeder_2015_14561]", "BiQ Analyzer[@bock2005biq]", "Methylation plotter[@mallona2014methylation]", "_MethVisual_[@zackay2010methvisual]", "_REDseq_[@zhu4redseq]", "_GenVisR_[@skidmore2016genvisr]")
inputs <- c("tab-delimited text", "JSON", "JSON", "BiQ methylation file", "tab-delimited text", "R list", "R list", "R dataframe")
online <- c("Yes", "Yes", "No", "Yes", "Yes", "No", "No", "No")
description <- c("interprets mutations with different heights along protein annotations in automatic color theme", "could combine different line and head colors with different drawing styles", "plots data point with different colors, heights, and size along annotations, and highlighted selcted coordinates", "interprets methylation status in black & white", "stacked multiple methylation status for multiple samples", "visualize the methylation status of CpGs according to their genomic position", "plot frequencies of methylations and SNPs along a chromosome", "plot most accurate graphic representation of the ensembl annotation version based on biomart service")
df <- data.frame(software=software, inputs=inputs, online=online, description=description)

library(pander)
panderOptions('table.alignment.default', function(df)
    ifelse(sapply(df, is.numeric), 'right', 'left'))
panderOptions('table.split.table', Inf)
panderOptions('keep.trailing.zeros', TRUE)
pander(df)
```


All the tools available for the genomic data visualization mentioned above can meet the basic even complicated requirement. However, if there are multiple mutations in almost same position, it is very difficult to display the data using any available tools. What's more there is a tendency that the figures it generated become more and more complex and busy. Thereby make it difficult in generating high quality pictures for publication in bunch. To fill this gap, we developed _trackViewer_, a R/Bioconductor package as an enhanced light-weight genome viewer for visualizing various types of high-throughput sequencing data, side by side or stacked with multiple annotation tracks. Besides the regular read coverage tracks supported by existing genome browsers/viewers,  _trackViewer_ can also be used to generate lollipop plot to depict the methylation and SNP/mutation status, together with coverage data and annotation tracks to facilitate integrated analysis of multi-omics data. In addition, figures generated by _trackViewer_ are interactive, i.e., the feel-and-look such as the layout, the color scheme and the order of tracks can be easily customized by the users. Furthermore, _trackViewer_ can be easily integrated into standard analysis pipeline for various high-throughput sequencing dataset such as ChIP-seq, RNA-seq, methylation-seq or DNA-seq. The images produced by _trackViewer_ are highly customizable including labels, symbols, colors and size. Here, we illustrate its utilities and capabilities in deriving biological insights from multi-omics dataset from GEO/Encode.


## Installation

_BiocManager_ is used to install the released version of _trackViewer_.

```{r eval=FALSE}
library(BiocManager)
install("trackViewer")
```

To install the development version of _trackViewer_, please try

```{r eval=FALSE}
library(BiocManager)
install("jianhong/trackViewer")
```

To check the current version of _trackViewer_, please try

```{r}
packageVersion("trackViewer")
```

## Quick start for lollipop plot

### Step 1. Prepare the methylation/variant/mutation data

The input data for lollipop plot by _trackViewer_ is an object of _GenomicRanges_::`GRanges`.

```{r quickStartStep1}
## load library
library(trackViewer)
set.seed(123) ## set seed for random sampling to make sure it can be repeat.
## Here we use SNP sample data
SNP <- c(10, 100, 105, 108, 400, 410, 420, 600, 700, 805, 840, 1400, 1402)
## use GenomicRanges::GRanges function to create a GRanges object.
## for real data, users can import vcf data via VariantAnnotation::readVcf function.
sample.gr <- GRanges("chr1", IRanges(SNP, width=1, 
                                     ## the name of GRanges object will be used as label
                                     names=paste0("snp", SNP)),
                     ## score value will be used to for the height of lollipop
                     score = sample.int(5, length(SNP), replace = TRUE),
                     ## set the color for lollipop node.
                     color = sample.int(6, length(SNP), replace = TRUE),
                     ## set the lollipop stem color
                     border = sample(c("black", "gray80", "gray30"),
                                     length(SNP), replace=TRUE)
                     )
```

### Step 2. Prepare the gene/protein model

The gene/protein model for lollipop plot by _trackViewer_ is also an object of _GenomicRanges_::`GRanges`.

```{r quickStartStep2}
features.gr <- GRanges("chr1", IRanges(c(1, 501, 1001), 
                                       width=c(120, 400, 405),
                                       names=paste0("exon", 1:3)),
                       fill = c("#FF8833", "#51C6E6", "#DFA32D"), ## color for exon
                       height = c(0.02, 0.05, 0.08) ## height for exon
                    )
```

### Step 3. Plot lollipop plot

The clustered events could be visualized one by one by jittered lollipop positions.

```{r quickStartStep3, fig.width=6, fig.height=3.5}
lolliplot(sample.gr, features.gr)
```


## Plot data from Variant Call Format (VCF) file

VCF is a text file format that contains metadata and mutation information about
genomic positions, original genotypes and optional genotypes. 
The _trackViewer_ package could show single nucleotide polymorphisms (SNPs) 
from VCF file in lollipop-style plot. Figure \@ref(fig:plotVCFdata) shows an 
example lollipop plot of real SNPs. Sample SNPs are a subset of 1000 variants 
and 50 samples from chromosome 22 taken from 1000 Genomes in VCF in the 
_VariantAnnotation_ package. 
Different colors depict the new SNP events in the circles. 
The number of circles indicates the number of SNP events. 

```{r plotVCFdata, fig.width=12,fig.height=5.5,dpi=72,fig.cap="lollipop plot for VCF data"}
library(VariantAnnotation) ## load package for reading vcf file
library(TxDb.Hsapiens.UCSC.hg19.knownGene) ## load package for gene model
library(org.Hs.eg.db) ## load package for gene name
library(rtracklayer)
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
## set the track range
gr <- GRanges("22", IRanges(50968014, 50970514, names="TYMP"))
## read in vcf file
tab <- TabixFile(fl)
vcf <- readVcf(fl, "hg19", param=gr)
## get GRanges from VCF object 
mutation.frequency <- rowRanges(vcf)
## keep the metadata
mcols(mutation.frequency) <- 
    cbind(mcols(mutation.frequency), 
          VariantAnnotation::info(vcf))
## set colors
mutation.frequency$border <- "gray30"
mutation.frequency$color <-
    ifelse(grepl("^rs", names(mutation.frequency)), 
           "lightcyan", "lavender")
## plot Global Allele Frequency based on AC/AN
mutation.frequency$score <- round(mutation.frequency$AF*100)
## change the SNPs label rotation angle
mutation.frequency$label.parameter.rot <- 45
## keep sequence level style same
seqlevelsStyle(gr) <- seqlevelsStyle(mutation.frequency) <- "UCSC"
seqlevels(mutation.frequency) <- "chr22"
## extract transcripts in the range
trs <- geneModelFromTxdb(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                         org.Hs.eg.db, gr=gr)
## subset the features to show the interested transcripts only
features <- c(range(trs[[1]]$dat), range(trs[[5]]$dat))
## define the feature labels
names(features) <- c(trs[[1]]$name, trs[[5]]$name)
## define the feature colors
features$fill <- c("lightblue", "mistyrose")
## define the feature heights
features$height <- c(.02, .04)
## set the legends
legends <- list(labels=c("known", "unkown"), 
                fill=c("lightcyan", "lavender"), 
                color=c("gray80", "gray80"))
## lollipop plot
lolliplot(mutation.frequency, 
          features, ranges=gr, type="circle", 
          legend=legends)
```

## Plot methylation data from bed format file

Any data can imported into a `GRanges` object can be viewed by _trackViewer_ package.
Sample methylations are random data generated for illustration and saved in bed format file.
The _rtracklayer_ package is used to import the methylation data into a `GRanges`.
The transcripts are extracted from `TxDb` object and assigned gene symbol with `org` database.
Here also show multiple transcripts could be shown in different colors and tracks.

```{r plotMethyBed, fig.width=12,fig.height=4,dpi=72,fig.cap="lollipop plot, pie layout"}
library(TxDb.Hsapiens.UCSC.hg19.knownGene) ## load package for gene model
library(org.Hs.eg.db) ## load package for gene name
library(rtracklayer)
## set the track range
gr <- GRanges("chr22", IRanges(50968014, 50970514, names="TYMP"))
## extract transcripts in the range
trs <- geneModelFromTxdb(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                         org.Hs.eg.db, gr=gr)
## subset the features to show the interested transcripts only
features <- GRangesList(trs[[1]]$dat, trs[[5]]$dat, trs[[6]]$dat)
flen <- elementNROWS(features)
features <- unlist(features)
## define the feature track layers
features$featureLayerID <- rep(1:2, c(sum(flen[-3]), flen[3]))
## define the feature labels
names(features) <- features$symbol
## define the feature colors
features$fill <- rep(c("lightblue", "mistyrose", "mistyrose"), flen)
## define the feature heights
features$height <- ifelse(features$feature=="CDS", .04, .02)
## import methylation data from a bed file
methy <- import(system.file("extdata", "methy.bed", package="trackViewer"), "BED")
## subset the data to simplify information
methy <- methy[methy$score > 20]
## for pie plot, there are must be at least two numeric columns
methy$score2 <- max(methy$score) - methy$score
## set the legends
legends <- list(labels=c("methylated", "unmethylated"), 
                fill=c("white", "lightblue"), 
                color=c("black", "black"))
## lollipop plot, pie layout
lolliplot(methy, features, 
          ranges=gr, type="pie", 
          legend=legends)
```

## Plot lollipop plot for multiple patients in "pie.stack" layout

The percentage of methylation rates are shown by pie graph in different layers 
for different patients. 

```{r plotMethypieStack, fig.width=12,fig.height=4,dpi=72,fig.cap="lollipop plot, pie.stack layout"}
## simulate multiple patients
rand.id <- sample.int(length(methy), 3*length(methy), replace=TRUE)
rand.id <- sort(rand.id)
methy.mul.patient <- methy[rand.id]
## pie.stack require metadata "stack.factor", and the metadata can not be 
## stack.factor.order or stack.factor.first
len.max <- max(table(rand.id))
stack.factors <- paste0("patient", 
                        formatC(1:len.max, 
                                width=nchar(as.character(len.max)), 
                                flag="0"))
methy.mul.patient$stack.factor <- 
    unlist(lapply(table(rand.id), sample, x=stack.factors))
methy.mul.patient$score <- 
    sample.int(100, length(methy.mul.patient), replace=TRUE)
## for a pie plot, two or more numeric meta-columns are required.
methy.mul.patient$score2 <- 100 - methy.mul.patient$score
## set different color set for different patient
patient.color.set <- as.list(as.data.frame(rbind(rainbow(length(stack.factors)), 
                                                 "#FFFFFFFF"), 
                                           stringsAsFactors=FALSE))
names(patient.color.set) <- stack.factors
methy.mul.patient$color <- 
    patient.color.set[methy.mul.patient$stack.factor]
## set the legends
legends <- list(labels=stack.factors, col="gray80", 
                fill=sapply(patient.color.set, `[`, 1))
## lollipop plot
lolliplot(methy.mul.patient, 
          features, ranges=gr, 
          type="pie.stack", 
          legend=legends)
```

## Plot lollipop plot in caterpillar layout to compare two samples

The caterpillar layout can be used to side by side comparing two samples or
to display dense data.

```{r plotCaterpillar, fig.width=12,fig.height=6,dpi=72,fig.cap="lollipop plot, caterpillar layout"}
## use SNPsideID to set the layer of event
sample.gr$SNPsideID <- sample(c("top", "bottom"), 
                               length(sample.gr),
                               replace=TRUE)
lolliplot(sample.gr, features.gr)
```

## Dandelion plot hundreds SNPs

Sometimes, there are as many as hundreds of SNPs or methylation status involved in one gene.
Dandelion plot can be used to depict such dense SNPs or methylation. 
Please note that the height of the dandelion indicates the density of the events.

```{r dandelionPlot, fig.width=8,fig.height=4,dpi=72,fig.cap="dandelion plot"}
methy <- import(system.file("extdata", "methy.bed", package="trackViewer"), "BED")
length(methy)
## set the color of dandelion leaves.
methy$color <- 3
methy$border <- "gray"
## we suppose the total event are same (methy + unmethy)
## we rescale the methylation score by max value of the score 
m <- max(methy$score)
methy$score <- methy$score/m
# The area of the fan indicate the percentage of methylation or rate of mutation.
dandelion.plot(methy, features, ranges=gr, type="fan")
```


