<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>IntegratedChIPseqWorkshop • integratedchipseqanalysisworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="IntegratedChIPseqWorkshop">
<meta property="og:description" content="integratedchipseqanalysisworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">integratedchipseqanalysisworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ChIPpeakAnno.html">IntegratedChIPseqWorkshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hukai916/IntegratedChIPseqWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>IntegratedChIPseqWorkshop</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/hukai916/IntegratedChIPseqWorkshop/blob/master/vignettes/ChIPpeakAnno.Rmd"><code>vignettes/ChIPpeakAnno.Rmd</code></a></small>
      <div class="hidden name"><code>ChIPpeakAnno.Rmd</code></div>

    </div>

    
    
<div id="load-required-libraries" class="section level2">
<h2 class="hasAnchor">
<a href="#load-required-libraries" class="anchor"></a>Load required libraries</h2>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"jianhong/workshop2020"</span>, <span class="kw">build_vignettes</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
</div>
<div id="four-steps-for-peak-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#four-steps-for-peak-annotation" class="anchor"></a>Four steps for peak annotation</h1>
<p>The functions, <code>toGRanges</code>, <code>annotatePeakInBatch</code>, and <code>addGeneIDs</code> in the <em>ChIPpeakAnno</em>, make the annotation of ChIP-Seq peaks streamlined into four major steps:</p>
<ol style="list-style-type: decimal">
<li><p>Read peak data with <code>toGRanges</code></p></li>
<li><p>Generate annotation data with <code>toGRanges</code></p></li>
<li><p>Annotate peaks with <code>annotatePeakInBatch</code></p></li>
<li><p>Add additional informations with <code>addGeneIDs</code></p></li>
</ol>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co">## First, load the ChIPpeakAnno package</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ChIPpeakAnno</span>)</pre></body></html></div>
<div id="step-1-convert-the-peak-data-from-bedbroadpeaknarrowpeak-etc--to-granges-with-the-togranges-method" class="section level2">
<h2 class="hasAnchor">
<a href="#step-1-convert-the-peak-data-from-bedbroadpeaknarrowpeak-etc--to-granges-with-the-togranges-method" class="anchor"></a>Step 1: convert the peak data from bed/broadPeak/narrowPeak etc. to <code>GRanges</code> with the <code>toGRanges</code> method</h2>
<p>Here we use broadPeak file format as example.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co">## the sample file is included in ChIPpeakAnno package. </span>
<span class="co">## chage the file path into your own file path to handle your data</span>
<span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"Tead4.broadPeak"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="co">## toGRanges is overlaoded method,</span>
<span class="co">## by define the correct file format to import the file in correct coordinates</span>
<span class="no">peaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">path</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"broadPeak"</span>)
<span class="co">## see top 2 lines of the imported peaks.</span>
<span class="co">## the imported peaks will be packaged into GRanges object</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">peaks</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 4 metadata columns:
##             seqnames        ranges strand |     score signalValue    pValue
##                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;   &lt;numeric&gt; &lt;numeric&gt;
##   peak12338     chr2 175473-176697      * |       206      668.42        -1
##   peak12339     chr2 246412-246950      * |        31      100.23        -1
##                qValue
##             &lt;numeric&gt;
##   peak12338        -1
##   peak12339        -1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="step-2-prepare-annotation-data-with-the-togranges-method" class="section level2">
<h2 class="hasAnchor">
<a href="#step-2-prepare-annotation-data-with-the-togranges-method" class="anchor"></a>Step 2: prepare annotation data with the <code>toGRanges</code> method</h2>
<p>ChIP-seq is a kind of DNA-seq. In the mapping step, no transcriptome is used. In the annotation step, both Ensembl or UCSC annotations can be used. Depend on your research, less complex genome annotations, such as UCSC annotations, are prefearable for reproducible and robust gene/transcripts annotation. To discover and explain unknown biological mechanisms, more comprehensive and complex genome annotations are necessary, such as Ensembl.</p>
<p>Here we show how to prepare annotations based on <strong>TxDb</strong> package with the <code>toGRanges</code> method.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)
<span class="no">annoDataTxDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoDataTxDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##      seqnames            ranges strand
##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##    1    chr19 58858172-58874214      -
##   10     chr8 18248755-18258723      +
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
<p>We can also prepare annotations based on <strong>EnsDb</strong> package with the <code>toGRanges</code> method.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">EnsDb.Hsapiens.v75</span>) <span class="co">##GRCh37.p13</span>
<span class="no">annoDataEnsDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">EnsDb.Hsapiens.v75</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoDataEnsDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##                   seqnames      ranges strand |   gene_name
##                      &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   ENSG00000223972     chr1 11869-14412      + |     DDX11L1
##   ENSG00000227232     chr1 14363-29806      - |      WASH7P
##   -------
##   seqinfo: 273 sequences from GRCh37 genome</code></pre>
<p>The shortage of offline annotation packages is that it may not contain up-to-date data. We can also use online annotation data via the <em>biomaRt</em> or <em>AnnotationHub</em> packages. However, to use the <em>biomaRt</em> package, you must be very clear that the default mart service is based the newrest genome assembly. For example, if you are annotating the data mapped to hg19/GRCh37 or mm9/NCBIM37 assembly, you must change the default server into achive server. For more details, please refer the documentation of the <em>biomaRt</em> package.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biomaRt</span>)
<span class="co">## here we use ensembl GRCh37.p13 (V75) service, </span>
<span class="co">## use listEnsemblArchives() function to list the available host</span>
<span class="co">## use listDatasets() function to list the available dataset</span>
<span class="no">mart</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span>(<span class="kw">biomart</span> <span class="kw">=</span> <span class="st">"ensembl"</span>, <span class="kw">dataset</span> <span class="kw">=</span> <span class="st">"hsapiens_gene_ensembl"</span>,
                <span class="kw">host</span> <span class="kw">=</span> <span class="st">"feb2014.archive.ensembl.org"</span>)
<span class="co">## use the getAnnotation function to obtain the TSS for ensembl GRCh37.p13.</span>
<span class="no">annoDataMart</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/getAnnotation.html">getAnnotation</a></span>(<span class="no">mart</span>, <span class="kw">featureType</span> <span class="kw">=</span> <span class="st">"TSS"</span>)
<span class="no">annoDataMart</span>[<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">annoDataEnsDb</span>), <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)]</pre></body></html></div>
<p>Here I show some sample code to get annotations from AnnotationHub.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">AnnotationHub</span>)
<span class="no">ah</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">AnnotationHub</a></span>()
<span class="no">hg19</span> <span class="kw">&lt;-</span> <span class="kw pkg">AnnotationHub</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span>(<span class="no">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"UCSC"</span>, <span class="st">"Hsapiens"</span>, <span class="st">"hg19"</span>, <span class="st">"knownGene"</span>))
<span class="no">hg19</span></pre></body></html></div>
<pre><code>## AnnotationHub with 1 record
## # snapshotDate(): 2020-04-27
## # names(): AH52258
## # $dataprovider: UCSC
## # $species: Homo sapiens
## # $rdataclass: TxDb
## # $rdatadateadded: 2016-12-22
## # $title: TxDb.Hsapiens.UCSC.hg19.knownGene.sqlite
## # $description: UCSC transcript based annotations for Homo sapiens
## # $taxonomyid: 9606
## # $genome: UCSC genomes
## # $sourcetype: FASTA
## # $sourceurl: http://genome.ucsc.edu/, http://hgdownload.cse.ucsc.edu/golden...
## # $sourcesize: NA
## # $tags: c("UCSC", "Transcript", "Annotation") 
## # retrieve record with 'object[["AH52258"]]'</code></pre>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">hg19TxDb</span> <span class="kw">&lt;-</span> <span class="no">hg19</span><span class="kw">[[</span><span class="st">"AH52258"</span>]]
<span class="no">annoDataHg19</span><span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">hg19TxDb</span>)
<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="no">annoDataHg19</span>, <span class="no">annoDataTxDb</span>)</pre></body></html></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co">## retrieve the gencode v32 annotation</span>
<span class="no">GRCh37</span> <span class="kw">&lt;-</span> <span class="kw pkg">AnnotationHub</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html">query</a></span>(<span class="no">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"GRCh37"</span>, <span class="st">"GENCODE"</span>, <span class="st">"v32"</span>, <span class="st">"TxDb"</span>))
<span class="no">annoDataGencode</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">GRCh37</span><span class="kw">[[</span><span class="st">"AH75188"</span>]])
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoDataGencode</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##                        seqnames            ranges strand
##                           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##   ENSG00000000003.15_3     chrX 99882105-99894988      -
##    ENSG00000000005.6_3     chrX 99839933-99854882      +
##   -------
##   seqinfo: 25 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="step-3-annotate-the-peaks-with-the-annotatepeakinbatch-function-" class="section level2">
<h2 class="hasAnchor">
<a href="#step-3-annotate-the-peaks-with-the-annotatepeakinbatch-function-" class="anchor"></a>Step 3: annotate the peaks with the <code>annotatePeakInBatch</code> function.</h2>
<p>Here we will compare the difference among differnt annotation sources.</p>
<p>First we compare the annotations from ensembl GRCh37.p13 between the <em>EnsDb.Hsapiens.v75</em> package and ensembl archives.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu">seqlevelsStyle</span>(<span class="no">annoDataEnsDb</span>)</pre></body></html></div>
<pre><code>## [1] "UCSC"</code></pre>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>)</pre></body></html></div>
<pre><code>## [1] "UCSC"</code></pre>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co">## keep the seqnames in the same style</span>
<span class="kw">if</span>(!<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>), <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataEnsDb</span>))){
  <span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>) <span class="kw">&lt;-</span> <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataEnsDb</span>)[<span class="fl">1</span>]
}
<span class="co">## do annotation by nearest TSS of Ensembl GRCh37.p13</span>
<span class="no">annoEnsDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span>(<span class="no">peaks</span>, <span class="kw">AnnotationData</span><span class="kw">=</span><span class="no">annoDataEnsDb</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoEnsDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 13 metadata columns:
##                             seqnames        ranges strand |     score
##                                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   peak12338.ENSG00000227061     chr2 175473-176697      * |       206
##   peak12339.ENSG00000143727     chr2 246412-246950      * |        31
##                             signalValue    pValue    qValue        peak
##                               &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;
##   peak12338.ENSG00000227061      668.42        -1        -1   peak12338
##   peak12339.ENSG00000143727      100.23        -1        -1   peak12339
##                                     feature start_position end_position
##                                 &lt;character&gt;      &lt;integer&gt;    &lt;integer&gt;
##   peak12338.ENSG00000227061 ENSG00000227061         197569       202605
##   peak12339.ENSG00000143727 ENSG00000143727         264140       278283
##                             feature_strand insideFeature distancetoFeature
##                                &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
##   peak12338.ENSG00000227061              +      upstream            -22096
##   peak12339.ENSG00000143727              +      upstream            -17728
##                             shortestDistance fromOverlappingOrNearest
##                                    &lt;integer&gt;              &lt;character&gt;
##   peak12338.ENSG00000227061            20872          NearestLocation
##   peak12339.ENSG00000143727            17190          NearestLocation
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="co"># A pie chart can be used to demonstrate the overlap features of the peaks.</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/pie1.html">pie1</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">annoEnsDb</span>$<span class="no">insideFeature</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/annotateEnsDb-1.png" alt="Figure 1. Pie chart of common peaks among EnsDb features." width="700"><p class="caption">
Figure 1. Pie chart of common peaks among EnsDb features.
</p>
</div>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="co">## keep the seqnames in the same style</span>
<span class="kw">if</span>(!<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>), <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataMart</span>))){
  <span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>) <span class="kw">&lt;-</span> <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataMart</span>)[<span class="fl">1</span>]
}
<span class="co">## do annotation by nearest TSS of Ensembl GRCh37.p13</span>
<span class="no">annoMart</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span>(<span class="no">peaks</span>, <span class="kw">AnnotationData</span><span class="kw">=</span><span class="no">annoDataMart</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoMart</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 13 metadata columns:
##                             seqnames        ranges strand |     score
##                                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   peak12338.ENSG00000227061     chr2 175473-176697      * |       206
##   peak12339.ENSG00000143727     chr2 246412-246950      * |        31
##                             signalValue    pValue    qValue        peak
##                               &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;
##   peak12338.ENSG00000227061      668.42        -1        -1   peak12338
##   peak12339.ENSG00000143727      100.23        -1        -1   peak12339
##                                     feature start_position end_position
##                                 &lt;character&gt;      &lt;integer&gt;    &lt;integer&gt;
##   peak12338.ENSG00000227061 ENSG00000227061         197569       202605
##   peak12339.ENSG00000143727 ENSG00000143727         264140       278283
##                             feature_strand insideFeature distancetoFeature
##                                &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
##   peak12338.ENSG00000227061              +      upstream            -22096
##   peak12339.ENSG00000143727              +      upstream            -17728
##                             shortestDistance fromOverlappingOrNearest
##                                    &lt;integer&gt;              &lt;character&gt;
##   peak12338.ENSG00000227061            20872          NearestLocation
##   peak12339.ENSG00000143727            17190          NearestLocation
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="co"># A pie chart can be used to demonstrate the overlap features of the peaks.</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/pie1.html">pie1</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">annoMart</span>$<span class="no">insideFeature</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/annotateBiomaRt-1.png" alt="Figure 2. Pie chart of common peaks among biomart features." width="700"><p class="caption">
Figure 2. Pie chart of common peaks among biomart features.
</p>
</div>
<p>We can see the minor difference between them. Lets check the reasons.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="co">## check how many annotations are different from each other</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">annoDataMart</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">annoDataEnsDb</span>)])<span class="kw">==</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">annoDataEnsDb</span>))</pre></body></html></div>
<pre><code>## 
## FALSE  TRUE 
##   417 63685</code></pre>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span>(<span class="no">annoDataMart</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">annoDataEnsDb</span>)])<span class="kw">==</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">end</a></span>(<span class="no">annoDataEnsDb</span>))</pre></body></html></div>
<pre><code>## 
## FALSE  TRUE 
##   417 63685</code></pre>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="co">## get the samples</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoDataMart</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">annoDataEnsDb</span>)][
  <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">annoDataMart</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">annoDataEnsDb</span>)])<span class="kw">!=</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">annoDataEnsDb</span>)], <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##           seqnames      ranges strand |
##              &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |
##   LRG_198  LRG_198  5001-40990      + |
##   LRG_262  LRG_262 33310-58918      + |
##                                                                                                            description
##                                                                                                            &lt;character&gt;
##   LRG_198                                                                           agrin [Source:HGNC Symbol;Acc:329]
##   LRG_262 pleckstrin homology domain containing, family G (with RhoGef domain) member 5 [Source:HGNC Symbol;Acc:29105]
##   -------
##   seqinfo: 690 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoDataEnsDb</span>[<span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">annoDataMart</span>[<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">annoDataEnsDb</span>)])<span class="kw">!=</span>
                     <span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span>(<span class="no">annoDataEnsDb</span>)], <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##           seqnames          ranges strand |   gene_name
##              &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   LRG_198     chr1   955503-991492      + |     LRG_198
##   LRG_262     chr1 6526152-6551760      - |     LRG_262
##   -------
##   seqinfo: 273 sequences from GRCh37 genome</code></pre>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="co">## check the biomart output</span>
<span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span>(<span class="kw">attributes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ensembl_gene_id"</span>, <span class="st">'chromosome_name'</span>, <span class="st">"description"</span>),
      <span class="kw">filters</span> <span class="kw">=</span> <span class="st">"ensembl_gene_id"</span>, <span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"LRG_198"</span>, <span class="st">"LRG_262"</span>),
      <span class="kw">mart</span> <span class="kw">=</span> <span class="no">mart</span>)</pre></body></html></div>
<pre><code>##   ensembl_gene_id chromosome_name
## 1         LRG_198         LRG_198
## 2         LRG_262         LRG_262
##                                                                                                    description
## 1                                                                           agrin [Source:HGNC Symbol;Acc:329]
## 2 pleckstrin homology domain containing, family G (with RhoGef domain) member 5 [Source:HGNC Symbol;Acc:29105]</code></pre>
<p>Here we can see that the <em>EnsDb.Hsapiens.v75</em> package already fixed the issue of chromosome_name issue, where the biomart online still output the wrong chromosome_name.</p>
<p>Now we annotate the peaks by UCSC hg19 annotations.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="co">## keep the seqnames in the same style</span>
<span class="kw">if</span>(!<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>), <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataTxDb</span>))){
  <span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>) <span class="kw">&lt;-</span> <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataTxDb</span>)[<span class="fl">1</span>]
}
<span class="co">## do annotation by nearest TSS of UCSC hg19 annotations</span>
<span class="no">annoTxDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span>(<span class="no">peaks</span>, <span class="kw">AnnotationData</span><span class="kw">=</span><span class="no">annoDataTxDb</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoTxDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 13 metadata columns:
##                   seqnames        ranges strand |     score signalValue
##                      &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;   &lt;numeric&gt;
##   peak12338.26751     chr2 175473-176697      * |       206      668.42
##   peak12339.26751     chr2 246412-246950      * |        31      100.23
##                      pValue    qValue        peak     feature start_position
##                   &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;      &lt;integer&gt;
##   peak12338.26751        -1        -1   peak12338       26751         218136
##   peak12339.26751        -1        -1   peak12339       26751         218136
##                   end_position feature_strand insideFeature distancetoFeature
##                      &lt;integer&gt;    &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
##   peak12338.26751       264810              -    downstream             89337
##   peak12339.26751       264810              -        inside             18398
##                   shortestDistance fromOverlappingOrNearest
##                          &lt;integer&gt;              &lt;character&gt;
##   peak12338.26751            41439          NearestLocation
##   peak12339.26751            17860          NearestLocation
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="co"># A pie chart can be used to demonstrate the overlap features of the peaks.</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/pie1.html">pie1</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">annoTxDb</span>$<span class="no">insideFeature</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/annotateTxDb-1.png" alt="Figure 3. Pie chart of common peaks among TxDb features." width="700"><p class="caption">
Figure 3. Pie chart of common peaks among TxDb features.
</p>
</div>
<p>Now we try to use gencode v32 annotations to annotate the peaks.</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="co">## keep the seqnames in the same style</span>
<span class="kw">if</span>(!<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>), <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataGencode</span>))){
  <span class="fu">seqlevelsStyle</span>(<span class="no">peaks</span>) <span class="kw">&lt;-</span> <span class="fu">seqlevelsStyle</span>(<span class="no">annoDataGencode</span>)[<span class="fl">1</span>]
}
<span class="co">## do annotation by nearest TSS of Gencode</span>
<span class="no">annoGencode</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span>(<span class="no">peaks</span>, <span class="kw">AnnotationData</span><span class="kw">=</span><span class="no">annoDataGencode</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoGencode</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 13 metadata columns:
##                                  seqnames        ranges strand |     score
##                                     &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##    peak12338.ENSG00000227061.2_6     chr2 175473-176697      * |       206
##   peak12339.ENSG00000143727.16_3     chr2 246412-246950      * |        31
##                                  signalValue    pValue    qValue        peak
##                                    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;
##    peak12338.ENSG00000227061.2_6      668.42        -1        -1   peak12338
##   peak12339.ENSG00000143727.16_3      100.23        -1        -1   peak12339
##                                               feature start_position
##                                           &lt;character&gt;      &lt;integer&gt;
##    peak12338.ENSG00000227061.2_6  ENSG00000227061.2_6         197569
##   peak12339.ENSG00000143727.16_3 ENSG00000143727.16_3         264140
##                                  end_position feature_strand insideFeature
##                                     &lt;integer&gt;    &lt;character&gt;   &lt;character&gt;
##    peak12338.ENSG00000227061.2_6       209892              +      upstream
##   peak12339.ENSG00000143727.16_3       278283              +      upstream
##                                  distancetoFeature shortestDistance
##                                          &lt;numeric&gt;        &lt;integer&gt;
##    peak12338.ENSG00000227061.2_6            -22096            20872
##   peak12339.ENSG00000143727.16_3            -17728            17190
##                                  fromOverlappingOrNearest
##                                               &lt;character&gt;
##    peak12338.ENSG00000227061.2_6          NearestLocation
##   peak12339.ENSG00000143727.16_3          NearestLocation
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="co"># A pie chart can be used to demonstrate the overlap features of the peaks.</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/pie1.html">pie1</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">annoGencode</span>$<span class="no">insideFeature</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/annotateGencode-1.png" alt="Figure 4. Pie chart of common peaks among Gencode features." width="700"><p class="caption">
Figure 4. Pie chart of common peaks among Gencode features.
</p>
</div>
<p>We can see the difference between the annotations among Ensembl, UCSC and Gencode in annotation. Let’s check the overlaps of annotations.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">Ensembl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">annoDataEnsDb</span>)
<span class="no">UCSC</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">annoDataTxDb</span>)
<span class="no">Gencode</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">annoDataGencode</span>)
<span class="co">## make the sequence info to same, otherwise can not compare.</span>
<span class="fu">genome</span>(<span class="fu">seqinfo</span>(<span class="no">Gencode</span>)) <span class="kw">&lt;-</span>
  <span class="fu">genome</span>(<span class="fu">seqinfo</span>(<span class="no">UCSC</span>)) <span class="kw">&lt;-</span>
  <span class="fu">genome</span>(<span class="fu">seqinfo</span>(<span class="no">Ensembl</span>)) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="st">"hg19"</span>)
<span class="fu">isCircular</span>(<span class="fu">seqinfo</span>(<span class="no">UCSC</span>)) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu">seqlevels</span>(<span class="no">UCSC</span>)))
<span class="fu">isCircular</span>(<span class="fu">seqinfo</span>(<span class="no">Gencode</span>)) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">FALSE</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="fu">seqlevels</span>(<span class="no">Gencode</span>)))
<span class="fu">seqlengths</span>(<span class="no">Ensembl</span>)[<span class="st">"chrM"</span>] <span class="kw">&lt;-</span> <span class="fu">seqlengths</span>(<span class="no">UCSC</span>)[<span class="st">"chrM"</span>]
<span class="co">## find the overlaps by max gap = 0 bp.</span>
<span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/findOverlapsOfPeaks.html">findOverlapsOfPeaks</a></span>(<span class="no">Ensembl</span>,
                          <span class="no">UCSC</span>,
                          <span class="no">Gencode</span>,
                          <span class="kw">ignore.strand</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
                          <span class="kw">connectedPeaks</span><span class="kw">=</span><span class="st">"keepAll"</span>)
<span class="co">## venn diagram to show the overlaps</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">connectedPeaks</span> <span class="kw">=</span> <span class="st">"keepAll"</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/Ensembl.vs.UCSC-1.png" alt="Figure 5. The overlap and intersection among UCSC, Ensembl, and Gencode annotations." width="700"><p class="caption">
Figure 5. The overlap and intersection among UCSC, Ensembl, and Gencode annotations.
</p>
</div>
<pre><code>## $p.value
##      Ensembl UCSC Gencode pval
## [1,]       0    1       1  NaN
## [2,]       1    0       1  NaN
## [3,]       1    1       0  NaN
## 
## $vennCounts
##      Ensembl UCSC Gencode Counts count.Ensembl count.UCSC count.Gencode
## [1,]       0    0       0      0             0          0             0
## [2,]       0    0       1   4016             0          0          4016
## [3,]       0    1       0    134             0        134             0
## [4,]       0    1       1     64             0         65            64
## [5,]       1    0       0   6182          6182          0             0
## [6,]       1    0       1  26961         27193          0         27373
## [7,]       1    1       0      9             9          9             0
## [8,]       1    1       1  22665         30652      22846         30744
## attr(,"class")
## [1] "VennCounts"</code></pre>
</div>
<div id="step-4-add-additional-annotation-with-the-addgeneids-function" class="section level2">
<h2 class="hasAnchor">
<a href="#step-4-add-additional-annotation-with-the-addgeneids-function" class="anchor"></a>Step 4: add additional annotation with the <code>addGeneIDs</code> function</h2>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>)
<span class="no">annoEnsDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/addGeneIDs.html">addGeneIDs</a></span>(<span class="no">annoEnsDb</span>, <span class="kw">orgAnn</span><span class="kw">=</span><span class="st">"org.Hs.eg.db"</span>,
                        <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"ensembl_gene_id"</span>,
                        <span class="kw">IDs2Add</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"symbol"</span>))
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoEnsDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 14 metadata columns:
##                             seqnames        ranges strand |     score
##                                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##   peak12338.ENSG00000227061     chr2 175473-176697      * |       206
##   peak12339.ENSG00000143727     chr2 246412-246950      * |        31
##                             signalValue    pValue    qValue        peak
##                               &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;
##   peak12338.ENSG00000227061      668.42        -1        -1   peak12338
##   peak12339.ENSG00000143727      100.23        -1        -1   peak12339
##                                     feature start_position end_position
##                                 &lt;character&gt;      &lt;integer&gt;    &lt;integer&gt;
##   peak12338.ENSG00000227061 ENSG00000227061         197569       202605
##   peak12339.ENSG00000143727 ENSG00000143727         264140       278283
##                             feature_strand insideFeature distancetoFeature
##                                &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
##   peak12338.ENSG00000227061              +      upstream            -22096
##   peak12339.ENSG00000143727              +      upstream            -17728
##                             shortestDistance fromOverlappingOrNearest
##                                    &lt;integer&gt;              &lt;character&gt;
##   peak12338.ENSG00000227061            20872          NearestLocation
##   peak12339.ENSG00000143727            17190          NearestLocation
##                                  symbol
##                             &lt;character&gt;
##   peak12338.ENSG00000227061        &lt;NA&gt;
##   peak12339.ENSG00000143727        ACP1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb46"><html><body><pre class="r"><span class="no">annoTxDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/addGeneIDs.html">addGeneIDs</a></span>(<span class="no">annoTxDb</span>, <span class="kw">orgAnn</span><span class="kw">=</span><span class="st">"org.Hs.eg.db"</span>,
                        <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>,
                        <span class="kw">IDs2Add</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"symbol"</span>))
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoTxDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 14 metadata columns:
##                   seqnames        ranges strand |     score signalValue
##                      &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;   &lt;numeric&gt;
##   peak12338.26751     chr2 175473-176697      * |       206      668.42
##   peak12339.26751     chr2 246412-246950      * |        31      100.23
##                      pValue    qValue        peak     feature start_position
##                   &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;      &lt;integer&gt;
##   peak12338.26751        -1        -1   peak12338       26751         218136
##   peak12339.26751        -1        -1   peak12339       26751         218136
##                   end_position feature_strand insideFeature distancetoFeature
##                      &lt;integer&gt;    &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
##   peak12338.26751       264810              -    downstream             89337
##   peak12339.26751       264810              -        inside             18398
##                   shortestDistance fromOverlappingOrNearest      symbol
##                          &lt;integer&gt;              &lt;character&gt; &lt;character&gt;
##   peak12338.26751            41439          NearestLocation      SH3YL1
##   peak12339.26751            17860          NearestLocation      SH3YL1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="no">annoGencode</span>$<span class="no">entrez_id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/xget.html">xget</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"(ENSG\\d{11}).*$"</span>, <span class="st">"\\1"</span>,
                                  <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">annoGencode</span>$<span class="no">feature</span>)),
                              <span class="no">org.Hs.egENSEMBL2EG</span>,
                              <span class="kw">output</span> <span class="kw">=</span> <span class="st">"last"</span>)
<span class="no">annoGencode</span>$<span class="no">symbol</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">annoGencode</span>$<span class="no">entrez_id</span>)] <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/xget.html">xget</a></span>(<span class="no">annoGencode</span>$<span class="no">entrez_id</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">annoGencode</span>$<span class="no">entrez_id</span>)], <span class="no">org.Hs.egSYMBOL</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoGencode</span>)</pre></body></html></div>
<pre><code>## GRanges object with 6 ranges and 15 metadata columns:
##                                  seqnames        ranges strand |     score
##                                     &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##    peak12338.ENSG00000227061.2_6     chr2 175473-176697      * |       206
##   peak12339.ENSG00000143727.16_3     chr2 246412-246950      * |        31
##   peak12340.ENSG00000143727.16_3     chr2 249352-250233      * |       195
##   peak12341.ENSG00000143727.16_3     chr2 259896-261404      * |       510
##   peak12342.ENSG00000143727.16_3     chr2 261931-263148      * |        48
##    peak12343.ENSG00000236856.1_6     chr2 378232-378871      * |       132
##                                  signalValue    pValue    qValue        peak
##                                    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt;
##    peak12338.ENSG00000227061.2_6      668.42        -1        -1   peak12338
##   peak12339.ENSG00000143727.16_3      100.23        -1        -1   peak12339
##   peak12340.ENSG00000143727.16_3      630.65        -1        -1   peak12340
##   peak12341.ENSG00000143727.16_3     1649.19        -1        -1   peak12341
##   peak12342.ENSG00000143727.16_3      155.56        -1        -1   peak12342
##    peak12343.ENSG00000236856.1_6      426.52        -1        -1   peak12343
##                                               feature start_position
##                                           &lt;character&gt;      &lt;integer&gt;
##    peak12338.ENSG00000227061.2_6  ENSG00000227061.2_6         197569
##   peak12339.ENSG00000143727.16_3 ENSG00000143727.16_3         264140
##   peak12340.ENSG00000143727.16_3 ENSG00000143727.16_3         264140
##   peak12341.ENSG00000143727.16_3 ENSG00000143727.16_3         264140
##   peak12342.ENSG00000143727.16_3 ENSG00000143727.16_3         264140
##    peak12343.ENSG00000236856.1_6  ENSG00000236856.1_6         388412
##                                  end_position feature_strand insideFeature
##                                     &lt;integer&gt;    &lt;character&gt;   &lt;character&gt;
##    peak12338.ENSG00000227061.2_6       209892              +      upstream
##   peak12339.ENSG00000143727.16_3       278283              +      upstream
##   peak12340.ENSG00000143727.16_3       278283              +      upstream
##   peak12341.ENSG00000143727.16_3       278283              +      upstream
##   peak12342.ENSG00000143727.16_3       278283              +      upstream
##    peak12343.ENSG00000236856.1_6       416885              +      upstream
##                                  distancetoFeature shortestDistance
##                                          &lt;numeric&gt;        &lt;integer&gt;
##    peak12338.ENSG00000227061.2_6            -22096            20872
##   peak12339.ENSG00000143727.16_3            -17728            17190
##   peak12340.ENSG00000143727.16_3            -14788            13907
##   peak12341.ENSG00000143727.16_3             -4244             2736
##   peak12342.ENSG00000143727.16_3             -2209              992
##    peak12343.ENSG00000236856.1_6            -10180             9541
##                                  fromOverlappingOrNearest   entrez_id
##                                               &lt;character&gt; &lt;character&gt;
##    peak12338.ENSG00000227061.2_6          NearestLocation        &lt;NA&gt;
##   peak12339.ENSG00000143727.16_3          NearestLocation          52
##   peak12340.ENSG00000143727.16_3          NearestLocation          52
##   peak12341.ENSG00000143727.16_3          NearestLocation          52
##   peak12342.ENSG00000143727.16_3          NearestLocation          52
##    peak12343.ENSG00000236856.1_6          NearestLocation        &lt;NA&gt;
##                                       symbol
##                                  &lt;character&gt;
##    peak12338.ENSG00000227061.2_6        &lt;NA&gt;
##   peak12339.ENSG00000143727.16_3        ACP1
##   peak12340.ENSG00000143727.16_3        ACP1
##   peak12341.ENSG00000143727.16_3        ACP1
##   peak12342.ENSG00000143727.16_3        ACP1
##    peak12343.ENSG00000236856.1_6        &lt;NA&gt;
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">UpSetR</span>)
<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/upset.html">upset</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/UpSetR/man/fromList.html">fromList</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">Ensembl</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">annoEnsDb</span>$<span class="no">symbol</span>),
                    <span class="kw">UCSC</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">annoTxDb</span>$<span class="no">symbol</span>),
                    <span class="kw">Gencode</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="no">annoGencode</span>$<span class="no">symbol</span>))),
      <span class="kw">order.by</span> <span class="kw">=</span> <span class="st">"freq"</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/addIDs-1.png" alt="Figure 6. upset plot of common gene symbols among annotations by Ensembl, UCSC and Gencode features." width="700"><p class="caption">
Figure 6. upset plot of common gene symbols among annotations by Ensembl, UCSC and Gencode features.
</p>
</div>
<p>As a conclusion, annotate with different annotation resources, even the other parameter keep same, the annotations will be different from each other. To improve the reproducibility, accuracy of an annotation source should be provided. Some people were advertizing their tools by playing this trick by using different annotations to mislead users.</p>
</div>
</div>
<div id="find-overlaps-for-replicates" class="section level1">
<h1 class="hasAnchor">
<a href="#find-overlaps-for-replicates" class="anchor"></a>Find overlaps for replicates</h1>
<p>The function <code>findOverlapsOfPeaks</code> returns an object of <strong>overlappingPeaks</strong>, which contains there elements: venn_cnt, peaklist (a list of<br>
overlapping peaks or unique peaks), and overlappingPeaks (a list of data frame consists of the annotation of all the overlapping peaks).</p>
<p>The following examples illustrate the usage of this method to convert BED and GFF file to GRanges, add metadata from orignal peaks to the overlap GRanges using function <code>addMetadata</code>, and visualize the overlapping using function <code>makeVennDiagram</code>.</p>
<div class="sourceCode" id="cb51"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ChIPpeakAnno</span>)
<span class="no">bed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"MACS_output.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="no">gr1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">bed</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"BED"</span>, <span class="kw">header</span><span class="kw">=</span><span class="fl">FALSE</span>)
<span class="co">## one can also try import from rtracklayer</span>
<span class="no">gff</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"GFF_peaks.gff"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="no">gr2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">gff</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"GFF"</span>, <span class="kw">header</span><span class="kw">=</span><span class="fl">FALSE</span>, <span class="kw">skip</span><span class="kw">=</span><span class="fl">3</span>)
<span class="co">## must keep the class exactly same as gr1$score, i.e., numeric.</span>
<span class="no">gr2</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="no">gr2</span>$<span class="no">score</span>)
<span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/findOverlapsOfPeaks.html">findOverlapsOfPeaks</a></span>(<span class="no">gr1</span>, <span class="no">gr2</span>, <span class="kw">connectedPeaks</span> <span class="kw">=</span> <span class="st">"keepAll"</span>)
<span class="co">## add metadata (mean of score) to the overlapping peaks</span>
<span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/addMetadata.html">addMetadata</a></span>(<span class="no">ol</span>, <span class="kw">colNames</span><span class="kw">=</span><span class="st">"score"</span>, <span class="kw">FUN</span><span class="kw">=</span><span class="no">mean</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">ol</span>$<span class="no">peaklist</span><span class="kw">[[</span><span class="st">"gr1///gr2"</span>]], <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames        ranges strand |                           peakNames
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |                     &lt;CharacterList&gt;
##   [1]     chr1 713791-715578      * | gr1__MACS_peak_13,gr2__001,gr2__002
##   [2]     chr1 724851-727191      * |          gr2__003,gr1__MACS_peak_14
##           score
##       &lt;numeric&gt;
##   [1]   850.203
##   [2]    29.170
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb53"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>), <span class="co"># circle fill color</span>
                <span class="kw">col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>), <span class="co">#circle border color</span>
                <span class="kw">cat.col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>)) <span class="co"># label color, keep same as circle border color</span></pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/findOverlapsOfPeaks-1.png" alt="Figure 7. Venn diagram of overlaps for replicated experiments" width="576"><p class="caption">
Figure 7. Venn diagram of overlaps for replicated experiments
</p>
</div>
<pre><code>## $p.value
##      gr1 gr2 pval
## [1,]   1   1    0
## 
## $vennCounts
##      gr1 gr2 Counts count.gr1 count.gr2
## [1,]   0   0      0         0         0
## [2,]   0   1     61         0        61
## [3,]   1   0     62        62         0
## [4,]   1   1    166       168       169
## attr(,"class")
## [1] "VennCounts"</code></pre>
<p>But if we check the sum of the overlaps, the total number for each sample are not identical to inputed peak number. This can be fixed by set ‘connectedPeaks’ parameter into ‘keepAll’.</p>
<div class="sourceCode" id="cb55"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">gr1</span>)</pre></body></html></div>
<pre><code>## [1] 230</code></pre>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">gr2</span>)</pre></body></html></div>
<pre><code>## [1] 230</code></pre>
<div class="sourceCode" id="cb59"><html><body><pre class="r"><span class="fl">62</span>+<span class="fl">166</span></pre></body></html></div>
<pre><code>## [1] 228</code></pre>
<div class="sourceCode" id="cb61"><html><body><pre class="r"><span class="fl">61</span>+<span class="fl">166</span></pre></body></html></div>
<pre><code>## [1] 227</code></pre>
<div class="sourceCode" id="cb63"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>),
                <span class="kw">col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>),
                <span class="kw">cat.col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>),
                <span class="kw">connectedPeaks</span> <span class="kw">=</span> <span class="st">"keepAll"</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/vennDiagramKeepAll-1.png" alt="Figure 8. Venn diagram of overlaps with all original counts." width="576"><p class="caption">
Figure 8. Venn diagram of overlaps with all original counts.
</p>
</div>
<pre><code>## $p.value
##      gr1 gr2 pval
## [1,]   1   1    0
## 
## $vennCounts
##      gr1 gr2 Counts count.gr1 count.gr2
## [1,]   0   0      0         0         0
## [2,]   0   1     61         0        61
## [3,]   1   0     62        62         0
## [4,]   1   1    166       168       169
## attr(,"class")
## [1] "VennCounts"</code></pre>
</div>
<div id="select-annotation-method" class="section level1">
<h1 class="hasAnchor">
<a href="#select-annotation-method" class="anchor"></a>Select annotation method</h1>
<div id="visualize-binding-site-distribution-relative-to-features" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-binding-site-distribution-relative-to-features" class="anchor"></a>Visualize binding site distribution relative to features</h2>
<p>The <code>assignChromosomeRegion</code> function can be used to summarize the distribution of peaks over different type of features such as exon, intron, enhancer, proximal promoter, 5’ UTR and 3’ UTR. This distribution can be summarized in peak centric or nucleotide centric view using the function <code>assignChromosomeRegion</code>. Please note that one peak might span multiple type of features, leading to the number of annotated features greater than the total number of input peaks. At the peak centric view, precedence will dictate the annotation order when peaks span multiple type of features.</p>
<p>The sample code here plots the distribution of peaks are enriched around the promoters.</p>
<div class="sourceCode" id="cb65"><html><body><pre class="r"><span class="no">overlaps</span> <span class="kw">&lt;-</span> <span class="no">ol</span>$<span class="no">peaklist</span><span class="kw">[[</span><span class="st">"gr1///gr2"</span>]] <span class="co">## get the overlapping peaks</span>
<span class="co">## load TxDb to assign genomic elements</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)
<span class="no">aCR</span><span class="kw">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/assignChromosomeRegion.html">assignChromosomeRegion</a></span>(<span class="no">overlaps</span>, <span class="kw">nucleotideLevel</span><span class="kw">=</span><span class="fl">FALSE</span>,
                           <span class="kw">precedence</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Promoters"</span>, <span class="st">"immediateDownstream"</span>,
                                         <span class="st">"fiveUTRs"</span>, <span class="st">"threeUTRs"</span>,
                                         <span class="st">"Exons"</span>, <span class="st">"Introns"</span>),
                           <span class="kw">TxDb</span><span class="kw">=</span><span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span>(<span class="no">aCR</span>$<span class="no">percentage</span>, <span class="kw">las</span><span class="kw">=</span><span class="fl">3</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/assignChromosomeRegion-1.png" alt="Figure 9. Peak distribution over different genomic features." width="960"><p class="caption">
Figure 9. Peak distribution over different genomic features.
</p>
</div>
<p>In addition, the distribution of the distance of overlapped peaks to the nearest feature such as the transcription start sites (TSS) can be plotted by the <code>binOverFeature</code> function. The sample code here plots the distribution of peaks around the TSS.</p>
<div class="sourceCode" id="cb66"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/binOverFeature.html">binOverFeature</a></span>(<span class="no">overlaps</span>, <span class="kw">annotationData</span><span class="kw">=</span><span class="no">annoDataTxDb</span>,
               <span class="kw">radius</span><span class="kw">=</span><span class="fl">5000</span>, <span class="kw">nbins</span><span class="kw">=</span><span class="fl">20</span>, <span class="kw">FUN</span><span class="kw">=</span><span class="no">length</span>, <span class="kw">errFun</span><span class="kw">=</span><span class="fl">0</span>,
               <span class="kw">ylab</span><span class="kw">=</span><span class="st">"count"</span>,
               <span class="kw">main</span><span class="kw">=</span><span class="st">"Distribution of aggregated peak numbers around TSS"</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/binOverFeature-1.png" alt="Figure 10. Distribution of peaks around transcript start sites." width="768"><p class="caption">
Figure 10. Distribution of peaks around transcript start sites.
</p>
</div>
</div>
<div id="annotate-peaks" class="section level2">
<h2 class="hasAnchor">
<a href="#annotate-peaks" class="anchor"></a>Annotate peaks</h2>
<p>The <code>annotatePeakInBatch</code> function provide multiple methods to annotate the peaks and those methods can be set by combination of multiple parameters. The ‘output’ is the key parameter to determine the annotation method. The default method is search the nearest features calculated as ‘PeakLoc - FeatureLocForDistance’. For more information, please refer the documentation of the <code>annotatePeakInBatch</code> function.</p>
<p>As shown from the distribution of aggregated peak numbers around TSS and the distribution of peaks in different of chromosome regions, most of the peaks locate around TSS. Therefore, it is reasonable to use the <code>annotatePeakInBatch</code> or <code>annoPeaks</code> to annotate the peaks to the promoter regions of Hg19/GRCh37 genes. Promoters can be specified with ‘bindingRegion’. For the following example, promoter region is defined as upstream 2000 and downstream 500 from TSS (bindingRegion=c(-2000, 500)).</p>
<div class="sourceCode" id="cb67"><html><body><pre class="r"><span class="no">overlaps.anno</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span>(<span class="no">overlaps</span>,
                                     <span class="kw">AnnotationData</span><span class="kw">=</span><span class="no">annoDataTxDb</span>,
                                     <span class="kw">output</span><span class="kw">=</span><span class="st">"nearestBiDirectionalPromoters"</span>,
                                     <span class="kw">bindingRegion</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">2000</span>, <span class="fl">500</span>))
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>)
<span class="no">overlaps.anno</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/addGeneIDs.html">addGeneIDs</a></span>(<span class="no">overlaps.anno</span>,
                            <span class="st">"org.Hs.eg.db"</span>,
                            <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>,
                            <span class="kw">IDs2Add</span> <span class="kw">=</span> <span class="st">"symbol"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">overlaps.anno</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 10 metadata columns:
##      seqnames        ranges strand |                           peakNames
##         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |                     &lt;CharacterList&gt;
##   X1     chr1 713791-715578      * | gr1__MACS_peak_13,gr2__001,gr2__002
##   X4     chr1 856361-856999      * |          gr1__MACS_peak_17,gr2__005
##          score        peak     feature feature.ranges feature.strand  distance
##      &lt;numeric&gt; &lt;character&gt; &lt;character&gt;      &lt;IRanges&gt;          &lt;Rle&gt; &lt;integer&gt;
##   X1   850.203          X1   100288069  700245-714068              -         0
##   X4    54.690          X4   100130417  852953-854817              -      1543
##      insideFeature distanceToSite       symbol
##        &lt;character&gt;      &lt;integer&gt;  &lt;character&gt;
##   X1  overlapStart              0 LOC100288069
##   X4      upstream           1543    LINC02593
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb69"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">WriteXLS</span>)
<span class="fu"><a href="https://rdrr.io/pkg/WriteXLS/man/WriteXLS.html">WriteXLS</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unname.html">unname</a></span>(<span class="no">overlaps.anno</span>)), <span class="st">"anno.xls"</span>)</pre></body></html></div>
<p>The distribution of the common peaks around features can be visualized using a pie chart.</p>
<div class="sourceCode" id="cb70"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/pie1.html">pie1</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">overlaps.anno</span>$<span class="no">insideFeature</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/pieChartBiPromoter-1.png" alt="Figure 11. Pie chart of the distribution of common peaks around features." width="700"><p class="caption">
Figure 11. Pie chart of the distribution of common peaks around features.
</p>
</div>
</div>
</div>
<div id="obtain-enriched-go-terms-and-pathways" class="section level1">
<h1 class="hasAnchor">
<a href="#obtain-enriched-go-terms-and-pathways" class="anchor"></a>Obtain enriched GO terms and Pathways</h1>
<p>The following example shows how to use <code>getEnrichedGO</code> to obtain a list of enriched GO terms with annotated peaks. For pathway analysis, please use function <code>getEnrichedPATH</code> with reactome or KEGG database. Please note that by default <em>feature_id_type</em> is set as “ensembl_gene_id”. If you are using <strong>TxDb</strong> as annotation data, please set it to “entrez_id”.</p>
<div class="sourceCode" id="cb71"><html><body><pre class="r"><span class="no">over</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/getEnrichedGO.html">getEnrichedGO</a></span>(<span class="no">overlaps.anno</span>, <span class="kw">orgAnn</span><span class="kw">=</span><span class="st">"org.Hs.eg.db"</span>,
                      <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>,
                      <span class="kw">maxP</span><span class="kw">=</span><span class="fl">.05</span>, <span class="kw">minGOterm</span><span class="kw">=</span><span class="fl">10</span>,
                      <span class="kw">multiAdjMethod</span><span class="kw">=</span><span class="st">"BH"</span>, <span class="kw">condense</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">over</span><span class="kw">[[</span><span class="st">"bp"</span>]][, -<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">10</span>)], <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## [1] go.id              go.term            Ontology           pvalue            
## [5] count.InDataset    count.InGenome     totaltermInDataset totaltermInGenome 
## [9] EntrezID          
## &lt;0 rows&gt; (or 0-length row.names)</code></pre>
<div class="sourceCode" id="cb73"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reactome.db</span>)
<span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/getEnrichedPATH.html">getEnrichedPATH</a></span>(<span class="no">overlaps.anno</span>, <span class="st">"org.Hs.eg.db"</span>, <span class="st">"reactome.db"</span>,
                        <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>, <span class="kw">maxP</span><span class="kw">=</span><span class="fl">.05</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">path</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>##        path.id EntrezID count.InDataset count.InGenome     pvalue
## 1 R-HSA-196849    55229               2            123 0.03029926
## 2 R-HSA-196849    65220               2            123 0.03029926
##   totaltermInDataset totaltermInGenome PATH
## 1                244            111075   NA
## 2                244            111075   NA</code></pre>
</div>
<div id="output-a-summary-of-consensus-in-the-peaks" class="section level1">
<h1 class="hasAnchor">
<a href="#output-a-summary-of-consensus-in-the-peaks" class="anchor"></a>Output a summary of consensus in the peaks</h1>
<p>There are multiple methods to get the consensus in the peaks:</p>
<ol style="list-style-type: decimal">
<li><p>output the fastq file by the <code>getAllPeakSequence</code> function and search the motif by the 3rd program such as homer, MEME, and so on.</p></li>
<li><p>test the pre-defined consensus patterns to see if target consensus are enriched or not by the <code>summarizePatternInPeaks</code> function.</p></li>
<li><p>calculate the z-scores of all combinations of oligonucleotide in a given length by Markove chain by the <code>oligoSummary</code> function.</p></li>
</ol>
<p>Here is an example to get the Z-scores for short oligos.</p>
<div class="sourceCode" id="cb75"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">seqinr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome.Hsapiens.UCSC.hg19</span>)
<span class="no">seq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/getAllPeakSequence.html">getAllPeakSequence</a></span>(<span class="no">overlaps</span>,
                          <span class="kw">upstream</span><span class="kw">=</span><span class="fl">20</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">20</span>,
                          <span class="kw">genome</span><span class="kw">=</span><span class="no">Hsapiens</span>)
<span class="co">## output the fasta file for the 3nd program</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/write2FASTA.html">write2FASTA</a></span>(<span class="no">seq</span>, <span class="st">"test.fa"</span>)
<span class="co">## summary of the short oligos</span>
<span class="no">os</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/oligoSummary.html">oligoSummary</a></span>(<span class="no">seq</span>, <span class="kw">oligoLength</span><span class="kw">=</span><span class="fl">6</span>, <span class="kw">MarkovOrder</span><span class="kw">=</span><span class="fl">3</span>,
                   <span class="kw">quickMotif</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="co">## plot the results</span>
<span class="no">zscore</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">os</span>$<span class="no">zscore</span>)
<span class="no">h</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="no">zscore</span>, <span class="kw">breaks</span><span class="kw">=</span><span class="fl">100</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"Histogram of Z-score"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="no">zscore</span>[<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">zscore</span>)], <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">h</span>$<span class="no">counts</span>)/<span class="fl">10</span>,
     <span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">zscore</span>[<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">zscore</span>)]), <span class="kw">srt</span><span class="kw">=</span><span class="fl">90</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/consensus-1.png" alt="Figure 12. Histogram of Z-score of 6-mer" width="576"><p class="caption">
Figure 12. Histogram of Z-score of 6-mer
</p>
</div>
<div class="sourceCode" id="cb76"><html><body><pre class="r"><span class="co">## generate the motifs</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">motifStack</span>)
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html">mapply</a></span>(<span class="kw">function</span>(<span class="no">.ele</span>, <span class="no">id</span>)
    <span class="fu">new</span>(<span class="st">"pfm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="no">.ele</span>, <span class="kw">name</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"SAMPLE motif"</span>, <span class="no">id</span>)),
    <span class="no">os</span>$<span class="no">motifs</span>, <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">os</span>$<span class="no">motifs</span>))
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifStack.html">motifStack</a></span>(<span class="no">pfms</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/simulation.motif-1.png" alt="Figure 13. Motif of simulation data" width="576"><p class="caption">
Figure 13. Motif of simulation data
</p>
</div>
</div>
<div id="find-peaks-with-bi-directional-promoters" class="section level1">
<h1 class="hasAnchor">
<a href="#find-peaks-with-bi-directional-promoters" class="anchor"></a>Find peaks with bi-directional promoters</h1>
<p>Bidirectional promoters are the DNA regions located between TSS of two adjacent genes that are transcribed on opposite directions and often co-regulated by this shared promoter region. Here is an example to find peaks near bi-directional promoters.</p>
<div class="sourceCode" id="cb77"><html><body><pre class="r"><span class="no">bdp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/peaksNearBDP.html">peaksNearBDP</a></span>(<span class="no">overlaps</span>, <span class="no">annoDataTxDb</span>, <span class="kw">maxgap</span><span class="kw">=</span><span class="fl">5000</span>)
<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">bdp</span>$<span class="no">percentPeaksWithBDP</span>,
  <span class="no">bdp</span>$<span class="no">n.peaks</span>,
  <span class="no">bdp</span>$<span class="no">n.peaksWithBDP</span>)</pre></body></html></div>
<pre><code>## [1]   0.05421687 166.00000000   9.00000000</code></pre>
<div class="sourceCode" id="cb79"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">bdp</span>$<span class="no">peaksWithBDP</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRangesList object of length 2:
## $`4`
## GRanges object with 2 ranges and 10 metadata columns:
##      seqnames        ranges strand |                  peakNames     score
##         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;CharacterList&gt; &lt;numeric&gt;
##   X4     chr1 856361-856999      * | gr1__MACS_peak_17,gr2__005     54.69
##   X4     chr1 856361-856999      * | gr1__MACS_peak_17,gr2__005     54.69
##        bdp_idx        peak     feature feature.ranges feature.strand  distance
##      &lt;integer&gt; &lt;character&gt; &lt;character&gt;      &lt;IRanges&gt;          &lt;Rle&gt; &lt;integer&gt;
##   X4         4          X4   100130417  852953-854817              -      1543
##   X4         4          X4      148398  860530-879961              +      3530
##      insideFeature distanceToSite
##        &lt;character&gt;      &lt;integer&gt;
##   X4      upstream           1543
##   X4      upstream           3530
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $`5`
## GRanges object with 2 ranges and 10 metadata columns:
##      seqnames        ranges strand |                  peakNames     score
##         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;CharacterList&gt; &lt;numeric&gt;
##   X5     chr1 859315-860144      * | gr2__006,gr1__MACS_peak_18    81.485
##   X5     chr1 859315-860144      * | gr2__006,gr1__MACS_peak_18    81.485
##        bdp_idx        peak     feature feature.ranges feature.strand  distance
##      &lt;integer&gt; &lt;character&gt; &lt;character&gt;      &lt;IRanges&gt;          &lt;Rle&gt; &lt;integer&gt;
##   X5         5          X5   100130417  852953-854817              -      4497
##   X5         5          X5      148398  860530-879961              +       385
##      insideFeature distanceToSite
##        &lt;character&gt;      &lt;integer&gt;
##   X5      upstream           4497
##   X5      upstream            385
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="determine-if-there-is-a-significant-overlap-among-multiple-sets-of-peaks" class="section level1">
<h1 class="hasAnchor">
<a href="#determine-if-there-is-a-significant-overlap-among-multiple-sets-of-peaks" class="anchor"></a>Determine if there is a significant overlap among multiple sets of peaks</h1>
<p>Given two or more peak lists from different TFs, one may be interested in finding whether DNA binding profile of those TFs are correlated, and if correlated, what is the common binding pattern.</p>
<p>Here we will show how to compare binding profiles from multiple transcription factors (TFs) by ChIP-seq sample data of TAF, YY1 and Tead4 from mouse.</p>
<div class="sourceCode" id="cb81"><html><body><pre class="r"><span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">path</span>, <span class="st">"broadPeak"</span>)
<span class="no">data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="no">files</span>), <span class="no">toGRanges</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"broadPeak"</span>)
(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">data</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".broadPeak"</span>, <span class="st">""</span>, <span class="no">files</span>))</pre></body></html></div>
<pre><code>## [1] "TAF"   "Tead4" "YY1"</code></pre>
<div id="hypergeometric-test" class="section level2">
<h2 class="hasAnchor">
<a href="#hypergeometric-test" class="anchor"></a>Hypergeometric test</h2>
<p>When we test the association between two sets of data based on hypergeometric distribution, the number of all potential binding sites is required. The parameter <em>totalTest</em> in the <code>makeVennDiagram</code> function indicates how many potential peaks in total will be used in the hypergeometric test. It should be larger than the largest number of peaks in the peak list. The smaller it is set, the more stringent the test is. The time used to calculate p-value does not depend on the value of the <em>totalTest</em>. For practical guidance on how to choose <em>totalTest</em>, please refer to the <a href="https://stat.ethz.ch/pipermail/bioconductor/2010-November/036540.html">post</a>. The following example makes an assumption that there are 3% of coding region plus promoter region. Because the sample data is only a subset of chromosome 2, we estimate that the total binding sites is 1/24 of possible binding region in the genome.</p>
<div class="sourceCode" id="cb83"><html><body><pre class="r"><span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/findOverlapsOfPeaks.html">findOverlapsOfPeaks</a></span>(<span class="no">data</span>, <span class="kw">connectedPeaks</span><span class="kw">=</span><span class="st">"keepAll"</span>)
<span class="no">averagePeakWidth</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="fu">width</span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu">GRangesList</span>(<span class="no">ol</span>$<span class="no">peaklist</span>))))
<span class="no">tot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span>(<span class="fl">3.3e+9</span> * <span class="fl">.03</span> / <span class="no">averagePeakWidth</span> / <span class="fl">24</span>)
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">totalTest</span><span class="kw">=</span><span class="no">tot</span>, <span class="kw">connectedPeaks</span><span class="kw">=</span><span class="st">"keepAll"</span>,
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#CC79A7"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#F0E442"</span>), <span class="co"># circle fill color</span>
                <span class="kw">col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>), <span class="co">#circle border color</span>
                <span class="kw">cat.col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/vennDiagram-1.png" alt="Figure 14. Venn diagram of overlaps." width="576"><p class="caption">
Figure 14. Venn diagram of overlaps.
</p>
</div>
<pre><code>## $p.value
##      TAF Tead4 YY1          pval
## [1,]   0     1   1  1.000000e+00
## [2,]   1     0   1 2.904297e-258
## [3,]   1     1   0  8.970986e-04
## 
## $vennCounts
##      TAF Tead4 YY1 Counts count.TAF count.Tead4 count.YY1
## [1,]   0     0   0    849         0           0         0
## [2,]   0     0   1    621         0           0       621
## [3,]   0     1   0   2097         0        2097         0
## [4,]   0     1   1    309         0         310       319
## [5,]   1     0   0     59        59           0         0
## [6,]   1     0   1    166       172           0       170
## [7,]   1     1   0      8         8           8         0
## [8,]   1     1   1    476       545         537       521
## attr(,"class")
## [1] "VennCounts"</code></pre>
<div class="sourceCode" id="cb85"><html><body><pre class="r"><span class="co">## see the difference if we set connectedPeaks to "keepFirstListConsistent"</span>
<span class="co">## set connectedPeaks to keepFirstListConsistent will show consistent total </span>
<span class="co">## number of peaks for the first peak list.</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">totalTest</span><span class="kw">=</span><span class="no">tot</span>, <span class="kw">connectedPeaks</span><span class="kw">=</span><span class="st">"keepFirstListConsistent"</span>,
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#CC79A7"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#F0E442"</span>),
                <span class="kw">col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>),
                <span class="kw">cat.col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/vennDiagramForFirstConsistent-1.png" alt="Figure 15. Venn diagram of overlaps for first TF." width="576"><p class="caption">
Figure 15. Venn diagram of overlaps for first TF.
</p>
</div>
<pre><code>## $p.value
##      TAF Tead4 YY1          pval
## [1,]   0     1   1  1.000000e+00
## [2,]   1     0   1 2.904297e-258
## [3,]   1     1   0  8.970986e-04
## 
## $vennCounts
##      TAF Tead4 YY1 Counts count.TAF count.Tead4 count.YY1
## [1,]   0     0   0    849         0           0         0
## [2,]   0     0   1    621         0           0       621
## [3,]   0     1   0   2097         0        2097         0
## [4,]   0     1   1    309         0         310       319
## [5,]   1     0   0     59        59           0         0
## [6,]   1     0   1    166       172           0       170
## [7,]   1     1   0      8         8           8         0
## [8,]   1     1   1    476       545         537       521
## attr(,"class")
## [1] "VennCounts"</code></pre>
</div>
<div id="permutation-test" class="section level2">
<h2 class="hasAnchor">
<a href="#permutation-test" class="anchor"></a>Permutation test</h2>
<p>The above hypergeometric test requires users to input an estimate of the total potential binding sites for a given TF. To circumvent this requirement, we implemented a permutation test called <code>peakPermTest</code>. Before performing a permutation test, users need to generate random peaks using the distribution discovered from the input peaks for a given feature type (transcripts or exons), to make sure the binding positions relative to features, such as TSS and geneEnd, and the width of the random peaks follow the distribution of that of the input peaks.</p>
<p>Alternatively, a peak pool representing all potential binding sites can be created with associated binding probabilities for random peak sampling using the <code>preparePool</code> function. Here is an example to build a peak pool for human genome using the transcription factor binding site clusters (V3) (see ?<code>wgEncodeTfbsV3</code>) downloaded from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeRegTfbsClusteredV3.bed.gz">ENCODE</a> with the HOT spots (?<code>HOT.spots</code>) removed. HOT spots are the genomic regions with high probability of being bound by many TFs in ChIP-seq experiments. We suggest remove those HOT spots from the peak lists before performing permutation test to avoid the overestimation of the association between the two input peak lists. Users can also choose to remove <a href="https://sites.google.com/site/anshulkundaje/projects/blacklists">ENCODE blacklist</a> for a given species. The blacklists were constructed by identifying consistently problematic regions over independent cell lines and types of experiments for each species in the ENCODE and modENCODE datasets. Please note that some of the blacklists may need to be converted to the correct genome assembly using liftover utility.</p>
<p>Following are the sample codes to do the permutation test using the <code>permTest</code> function:</p>
<div class="sourceCode" id="cb87"><html><body><pre class="r">    <span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">HOT.spots</span>)
    <span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">wgEncodeTfbsV3</span>)
    <span class="no">hotGR</span> <span class="kw">&lt;-</span> <span class="fu">reduce</span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">HOT.spots</span>))
    <span class="no">removeOl</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">.ele</span>){
        <span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu">findOverlaps</span>(<span class="no">.ele</span>, <span class="no">hotGR</span>)
        <span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">ol</span>)<span class="kw">&gt;</span><span class="fl">0</span>) <span class="no">.ele</span> <span class="kw">&lt;-</span> <span class="no">.ele</span>[-<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu">queryHits</span>(<span class="no">ol</span>))]
        <span class="no">.ele</span>
    }
    <span class="no">TAF</span> <span class="kw">&lt;-</span> <span class="fu">removeOl</span>(<span class="no">data</span><span class="kw">[[</span><span class="st">"TAF"</span>]])
    <span class="no">TEAD4</span> <span class="kw">&lt;-</span> <span class="fu">removeOl</span>(<span class="no">data</span><span class="kw">[[</span><span class="st">"Tead4"</span>]])
    <span class="no">YY1</span> <span class="kw">&lt;-</span> <span class="fu">removeOl</span>(<span class="no">data</span><span class="kw">[[</span><span class="st">"YY1"</span>]])
    <span class="co"># we subset the pool to save demo time</span>
    <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
    <span class="no">wgEncodeTfbsV3.subset</span> <span class="kw">&lt;-</span>
        <span class="no">wgEncodeTfbsV3</span>[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">wgEncodeTfbsV3</span>), <span class="fl">2000</span>)]
    <span class="no">pool</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"permPool"</span>, <span class="kw">grs</span><span class="kw">=</span><span class="fu">GRangesList</span>(<span class="no">wgEncodeTfbsV3.subset</span>), <span class="kw">N</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">YY1</span>))
    <span class="no">pt1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/peakPermTest.html">peakPermTest</a></span>(<span class="no">YY1</span>, <span class="no">TEAD4</span>, <span class="kw">pool</span><span class="kw">=</span><span class="no">pool</span>, <span class="kw">seed</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">force.parallel</span><span class="kw">=</span><span class="fl">FALSE</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">pt1</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/peakPermTest1-1.png" alt="Figure 16. permutation test for YY1 and TEAD4" width="700"><p class="caption">
Figure 16. permutation test for YY1 and TEAD4
</p>
</div>
<div class="sourceCode" id="cb88"><html><body><pre class="r">    <span class="no">pt2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/peakPermTest.html">peakPermTest</a></span>(<span class="no">YY1</span>, <span class="no">TAF</span>, <span class="kw">pool</span><span class="kw">=</span><span class="no">pool</span>, <span class="kw">seed</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">force.parallel</span><span class="kw">=</span><span class="fl">FALSE</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">pt2</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/peakPermTest2-1.png" alt="Figure 17. permutation test for YY1 and TAF" width="700"><p class="caption">
Figure 17. permutation test for YY1 and TAF
</p>
</div>
</div>
</div>
<div id="metagene-analysis-for-given-featurepeak-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#metagene-analysis-for-given-featurepeak-ranges" class="anchor"></a>Metagene analysis for given feature/peak ranges</h1>
<p>You can easily visualize and compare the binding patterns of raw signals of multiple ChIP-Seq experiments using function <code>featureAlignedHeatmap</code> and <code>featureAlignedDistribution</code>.</p>
<div class="sourceCode" id="cb89"><html><body><pre class="r"><span class="no">features</span> <span class="kw">&lt;-</span> <span class="no">ol</span>$<span class="no">peaklist</span><span class="kw">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">ol</span>$<span class="no">peaklist</span>)]]
<span class="no">feature.recentered</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html">reCenterPeaks</a></span>(<span class="no">features</span>, <span class="kw">width</span><span class="kw">=</span><span class="fl">4000</span>)
<span class="co">## here we also suggest importData function in bioconductor trackViewer package </span>
<span class="co">## to import the coverage.</span>
<span class="co">## compare rtracklayer, it will save you time when handle huge dataset.</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">path</span>, <span class="st">"bigWig"</span>)
<span class="kw">if</span>(<span class="no">.Platform</span>$<span class="no">OS.type</span> <span class="kw">!=</span> <span class="st">"windows"</span>){
    <span class="no">cvglists</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="no">files</span>), <span class="no">import</span>,
                       <span class="kw">format</span><span class="kw">=</span><span class="st">"BigWig"</span>,
                       <span class="kw">which</span><span class="kw">=</span><span class="no">feature.recentered</span>,
                       <span class="kw">as</span><span class="kw">=</span><span class="st">"RleList"</span>)
}<span class="kw">else</span>{<span class="co">## rtracklayer can not import bigWig files on Windows</span>
    <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"cvglist.rds"</span>))
}
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">cvglists</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".bigWig"</span>, <span class="st">""</span>, <span class="no">files</span>)
<span class="no">feature.center</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html">reCenterPeaks</a></span>(<span class="no">features</span>, <span class="kw">width</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">sig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedSignal.html">featureAlignedSignal</a></span>(<span class="no">cvglists</span>, <span class="no">feature.center</span>,
                            <span class="kw">upstream</span><span class="kw">=</span><span class="fl">2000</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">2000</span>)
<span class="co">##Because the bw file is only a subset of the original file,</span>
<span class="co">##the signals are not exists for every peak.</span>
<span class="no">keep</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">sig</span><span class="kw">[[</span><span class="fl">2</span>]]) <span class="kw">&gt;</span> <span class="fl">0</span>
<span class="no">sig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">sig</span>, <span class="kw">function</span>(<span class="no">.ele</span>) <span class="no">.ele</span>[<span class="no">keep</span>, ], <span class="kw">simplify</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">feature.center</span> <span class="kw">&lt;-</span> <span class="no">feature.center</span>[<span class="no">keep</span>]
<span class="no">heatmap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedHeatmap.html">featureAlignedHeatmap</a></span>(<span class="no">sig</span>, <span class="no">feature.center</span>,
                                 <span class="kw">upstream</span><span class="kw">=</span><span class="fl">2000</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">2000</span>,
                                 <span class="kw">upper.extreme</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>,<span class="fl">.5</span>,<span class="fl">4</span>))</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/heatmap-1.png" alt="Figure 18. Heatmap of aligned features sorted by signal of TAF" width="384"><p class="caption">
Figure 18. Heatmap of aligned features sorted by signal of TAF
</p>
</div>
<div class="sourceCode" id="cb90"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedDistribution.html">featureAlignedDistribution</a></span>(<span class="no">sig</span>, <span class="no">feature.center</span>,
                           <span class="kw">upstream</span><span class="kw">=</span><span class="fl">2000</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">2000</span>,
                           <span class="kw">type</span><span class="kw">=</span><span class="st">"l"</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/distribution-1.png" alt="Figure 19. Distribution of aligned features" width="576"><p class="caption">
Figure 19. Distribution of aligned features
</p>
</div>
</div>
<div id="below-is-for-trackviewer" class="section level1">
<h1 class="hasAnchor">
<a href="#below-is-for-trackviewer" class="anchor"></a>Below is for trackViewer:</h1>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>It is a routine way to show the distribution of mutation for genetic variations by lollipop-style (or needle-style) plot in a genome browser, along with a variety of genomic annotations, such as gene level or exon level models, CpG island, and so on. To show the SNP status or methylation data, a special plot, called lollipop plot or needle plot, is used to show the distribution along annotations. Many of currently available tools offline or online provide lollipop-like plot to show the mutation data such as cBioPortal Tools::MutationMapper<span class="citation">[@gao2013integrative]</span>, EMBL-EBI::Pfam<span class="citation">[@finn2006pfam]</span>, and BioJS::muts-needle-plot<span class="citation">[@michael_p_schroeder_2015_14561]</span>, BiQ Analyzer<span class="citation">[@bock2005biq]</span>, and Methylation plotter<span class="citation">[@mallona2014methylation]</span>. The cBioPortal Tools::MutationMapper is a well-known and easy to use on-line genome browser that can output high quality of figures with mutations by input tab-delimited mutation data. In R/Bioconductor, there are options to use the flexibility of the Rgraphics system to display Methylation data such as <em>MethVisual</em><span class="citation">[@zackay2010methvisual]</span>, <em>REDseq</em><span class="citation">[@zhu4redseq]</span>, and <em>GenVisR</em><span class="citation">[@skidmore2016genvisr]</span>.</p>
<p><strong>Table 1: Tools availble for lollipop plot. </strong></p>
<table class="table">
<colgroup>
<col width="44%">
<col width="19%">
<col width="7%">
<col width="28%">
</colgroup>
<thead><tr class="header">
<th align="left">software</th>
<th align="left">inputs</th>
<th align="left">online</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MutationMapper<span class="citation">[@gao2013integrative]</span>
</td>
<td align="left">tab-delimited text</td>
<td align="left">Yes</td>
<td align="left">interprets mutations with different heights along protein annotations in automatic color theme</td>
</tr>
<tr class="even">
<td align="left">Pfam<span class="citation">[@finn2006pfam]</span>
</td>
<td align="left">JSON</td>
<td align="left">Yes</td>
<td align="left">could combine different line and head colors with different drawing styles</td>
</tr>
<tr class="odd">
<td align="left">muts-needle-plot<span class="citation">[@michael_p_schroeder_2015_14561]</span>
</td>
<td align="left">JSON</td>
<td align="left">No</td>
<td align="left">plots data point with different colors, heights, and size along annotations, and highlighted selcted coordinates</td>
</tr>
<tr class="even">
<td align="left">BiQ Analyzer<span class="citation">[@bock2005biq]</span>
</td>
<td align="left">BiQ methylation file</td>
<td align="left">Yes</td>
<td align="left">interprets methylation status in black &amp; white</td>
</tr>
<tr class="odd">
<td align="left">Methylation plotter<span class="citation">[@mallona2014methylation]</span>
</td>
<td align="left">tab-delimited text</td>
<td align="left">Yes</td>
<td align="left">stacked multiple methylation status for multiple samples</td>
</tr>
<tr class="even">
<td align="left">
<em>MethVisual</em><span class="citation">[@zackay2010methvisual]</span>
</td>
<td align="left">R list</td>
<td align="left">No</td>
<td align="left">visualize the methylation status of CpGs according to their genomic position</td>
</tr>
<tr class="odd">
<td align="left">
<em>REDseq</em><span class="citation">[@zhu4redseq]</span>
</td>
<td align="left">R list</td>
<td align="left">No</td>
<td align="left">plot frequencies of methylations and SNPs along a chromosome</td>
</tr>
<tr class="even">
<td align="left">
<em>GenVisR</em><span class="citation">[@skidmore2016genvisr]</span>
</td>
<td align="left">R dataframe</td>
<td align="left">No</td>
<td align="left">plot most accurate graphic representation of the ensembl annotation version based on biomart service</td>
</tr>
</tbody>
</table>
<p>All the tools available for the genomic data visualization mentioned above can meet the basic even complicated requirement. However, if there are multiple mutations in almost same position, it is very difficult to display the data using any available tools. What’s more there is a tendency that the figures it generated become more and more complex and busy. Thereby make it difficult in generating high quality pictures for publication in bunch. To fill this gap, we developed <em>trackViewer</em>, a R/Bioconductor package as an enhanced light-weight genome viewer for visualizing various types of high-throughput sequencing data, side by side or stacked with multiple annotation tracks. Besides the regular read coverage tracks supported by existing genome browsers/viewers, <em>trackViewer</em> can also be used to generate lollipop plot to depict the methylation and SNP/mutation status, together with coverage data and annotation tracks to facilitate integrated analysis of multi-omics data. In addition, figures generated by <em>trackViewer</em> are interactive, i.e., the feel-and-look such as the layout, the color scheme and the order of tracks can be easily customized by the users. Furthermore, <em>trackViewer</em> can be easily integrated into standard analysis pipeline for various high-throughput sequencing dataset such as ChIP-seq, RNA-seq, methylation-seq or DNA-seq. The images produced by <em>trackViewer</em> are highly customizable including labels, symbols, colors and size. Here, we illustrate its utilities and capabilities in deriving biological insights from multi-omics dataset from GEO/Encode.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><em>BiocManager</em> is used to install the released version of <em>trackViewer</em>.</p>
<div class="sourceCode" id="cb91"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)
<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"trackViewer"</span>)</pre></body></html></div>
<p>To install the development version of <em>trackViewer</em>, please try</p>
<div class="sourceCode" id="cb92"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)
<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"jianhong/trackViewer"</span>)</pre></body></html></div>
<p>To check the current version of <em>trackViewer</em>, please try</p>
<div class="sourceCode" id="cb93"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span>(<span class="st">"trackViewer"</span>)</pre></body></html></div>
<pre><code>## [1] '1.24.1'</code></pre>
</div>
<div id="quick-start-for-lollipop-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start-for-lollipop-plot" class="anchor"></a>Quick start for lollipop plot</h2>
<div id="step-1--prepare-the-methylationvariantmutation-data" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1--prepare-the-methylationvariantmutation-data" class="anchor"></a>Step 1. Prepare the methylation/variant/mutation data</h3>
<p>The input data for lollipop plot by <em>trackViewer</em> is an object of <em>GenomicRanges</em>::<code>GRanges</code>.</p>
<div class="sourceCode" id="cb95"><html><body><pre class="r"><span class="co">## load library</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">trackViewer</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## set seed for random sampling to make sure it can be repeat.</span>
<span class="co">## Here we use SNP sample data</span>
<span class="no">SNP</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">10</span>, <span class="fl">100</span>, <span class="fl">105</span>, <span class="fl">108</span>, <span class="fl">400</span>, <span class="fl">410</span>, <span class="fl">420</span>, <span class="fl">600</span>, <span class="fl">700</span>, <span class="fl">805</span>, <span class="fl">840</span>, <span class="fl">1400</span>, <span class="fl">1402</span>)
<span class="co">## use GenomicRanges::GRanges function to create a GRanges object.</span>
<span class="co">## for real data, users can import vcf data via VariantAnnotation::readVcf function.</span>
<span class="no">sample.gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr1"</span>, <span class="fu">IRanges</span>(<span class="no">SNP</span>, <span class="kw">width</span><span class="kw">=</span><span class="fl">1</span>,
                                     <span class="co">## the name of GRanges object will be used as label</span>
                                     <span class="kw">names</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"snp"</span>, <span class="no">SNP</span>)),
                     <span class="co">## score value will be used to for the height of lollipop</span>
                     <span class="kw">score</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                     <span class="co">## set the color for lollipop node.</span>
                     <span class="kw">color</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">6</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                     <span class="co">## set the lollipop stem color</span>
                     <span class="kw">border</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"gray80"</span>, <span class="st">"gray30"</span>),
                                     <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
                     )</pre></body></html></div>
</div>
<div id="step-2--prepare-the-geneprotein-model" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2--prepare-the-geneprotein-model" class="anchor"></a>Step 2. Prepare the gene/protein model</h3>
<p>The gene/protein model for lollipop plot by <em>trackViewer</em> is also an object of <em>GenomicRanges</em>::<code>GRanges</code>.</p>
<div class="sourceCode" id="cb96"><html><body><pre class="r"><span class="no">features.gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr1"</span>, <span class="fu">IRanges</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">501</span>, <span class="fl">1001</span>),
                                       <span class="kw">width</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">120</span>, <span class="fl">400</span>, <span class="fl">405</span>),
                                       <span class="kw">names</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"exon"</span>, <span class="fl">1</span>:<span class="fl">3</span>)),
                       <span class="kw">fill</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#FF8833"</span>, <span class="st">"#51C6E6"</span>, <span class="st">"#DFA32D"</span>), <span class="co">## color for exon</span>
                       <span class="kw">height</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.02</span>, <span class="fl">0.05</span>, <span class="fl">0.08</span>) <span class="co">## height for exon</span>
                    )</pre></body></html></div>
</div>
<div id="step-3--plot-lollipop-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3--plot-lollipop-plot" class="anchor"></a>Step 3. Plot lollipop plot</h3>
<p>The clustered events could be visualized one by one by jittered lollipop positions.</p>
<div class="sourceCode" id="cb97"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">sample.gr</span>, <span class="no">features.gr</span>)</pre></body></html></div>
<p><img src="ChIPpeakAnno_files/figure-html/quickStartStep3-1.png" width="576"></p>
</div>
</div>
<div id="plot-data-from-variant-call-format-vcf-file" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-data-from-variant-call-format-vcf-file" class="anchor"></a>Plot data from Variant Call Format (VCF) file</h2>
<p>VCF is a text file format that contains metadata and mutation information about genomic positions, original genotypes and optional genotypes. The <em>trackViewer</em> package could show single nucleotide polymorphisms (SNPs) from VCF file in lollipop-style plot. Figure @ref(fig:plotVCFdata) shows an example lollipop plot of real SNPs. Sample SNPs are a subset of 1000 variants and 50 samples from chromosome 22 taken from 1000 Genomes in VCF in the <em>VariantAnnotation</em> package. Different colors depict the new SNP events in the circles. The number of circles indicates the number of SNP events.</p>
<div class="sourceCode" id="cb98"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">VariantAnnotation</span>) <span class="co">## load package for reading vcf file</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>) <span class="co">## load package for gene model</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>) <span class="co">## load package for gene name</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)
<span class="no">fl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"chr22.vcf.gz"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"VariantAnnotation"</span>)
<span class="co">## set the track range</span>
<span class="no">gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"22"</span>, <span class="fu">IRanges</span>(<span class="fl">50968014</span>, <span class="fl">50970514</span>, <span class="kw">names</span><span class="kw">=</span><span class="st">"TYMP"</span>))
<span class="co">## read in vcf file</span>
<span class="no">tab</span> <span class="kw">&lt;-</span> <span class="fu">TabixFile</span>(<span class="no">fl</span>)
<span class="no">vcf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(<span class="no">fl</span>, <span class="st">"hg19"</span>, <span class="kw">param</span><span class="kw">=</span><span class="no">gr</span>)
<span class="co">## get GRanges from VCF object </span>
<span class="no">mutation.frequency</span> <span class="kw">&lt;-</span> <span class="fu">rowRanges</span>(<span class="no">vcf</span>)
<span class="co">## keep the metadata</span>
<span class="fu">mcols</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="fu">mcols</span>(<span class="no">mutation.frequency</span>),
          <span class="kw pkg">VariantAnnotation</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html">info</a></span>(<span class="no">vcf</span>))
<span class="co">## set colors</span>
<span class="no">mutation.frequency</span>$<span class="no">border</span> <span class="kw">&lt;-</span> <span class="st">"gray30"</span>
<span class="no">mutation.frequency</span>$<span class="no">color</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"^rs"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">mutation.frequency</span>)),
           <span class="st">"lightcyan"</span>, <span class="st">"lavender"</span>)
<span class="co">## plot Global Allele Frequency based on AC/AN</span>
<span class="no">mutation.frequency</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="no">mutation.frequency</span>$<span class="no">AF</span>*<span class="fl">100</span>)
<span class="co">## change the SNPs label rotation angle</span>
<span class="no">mutation.frequency</span>$<span class="no">label.parameter.rot</span> <span class="kw">&lt;-</span> <span class="fl">45</span>
<span class="co">## keep sequence level style same</span>
<span class="fu">seqlevelsStyle</span>(<span class="no">gr</span>) <span class="kw">&lt;-</span> <span class="fu">seqlevelsStyle</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span> <span class="st">"UCSC"</span>
<span class="fu">seqlevels</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span> <span class="st">"chr22"</span>
<span class="co">## extract transcripts in the range</span>
<span class="no">trs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/geneModelFromTxdb.html">geneModelFromTxdb</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="no">org.Hs.eg.db</span>, <span class="kw">gr</span><span class="kw">=</span><span class="no">gr</span>)
<span class="co">## subset the features to show the interested transcripts only</span>
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">dat</span>), <span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">dat</span>))
<span class="co">## define the feature labels</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">features</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">name</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">name</span>)
<span class="co">## define the feature colors</span>
<span class="no">features</span>$<span class="no">fill</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightblue"</span>, <span class="st">"mistyrose"</span>)
<span class="co">## define the feature heights</span>
<span class="no">features</span>$<span class="no">height</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">.02</span>, <span class="fl">.04</span>)
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"known"</span>, <span class="st">"unkown"</span>),
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightcyan"</span>, <span class="st">"lavender"</span>),
                <span class="kw">color</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gray80"</span>, <span class="st">"gray80"</span>))
<span class="co">## lollipop plot</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">mutation.frequency</span>,
          <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"circle"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/plotVCFdata-1.png" alt="lollipop plot for VCF data" width="864"><p class="caption">
lollipop plot for VCF data
</p>
</div>
</div>
<div id="plot-methylation-data-from-bed-format-file" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-methylation-data-from-bed-format-file" class="anchor"></a>Plot methylation data from bed format file</h2>
<p>Any data can imported into a <code>GRanges</code> object can be viewed by <em>trackViewer</em> package. Sample methylations are random data generated for illustration and saved in bed format file. The <em>rtracklayer</em> package is used to import the methylation data into a <code>GRanges</code>. The transcripts are extracted from <code>TxDb</code> object and assigned gene symbol with <code>org</code> database. Here also show multiple transcripts could be shown in different colors and tracks.</p>
<div class="sourceCode" id="cb99"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>) <span class="co">## load package for gene model</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>) <span class="co">## load package for gene name</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)
<span class="co">## set the track range</span>
<span class="no">gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr22"</span>, <span class="fu">IRanges</span>(<span class="fl">50968014</span>, <span class="fl">50970514</span>, <span class="kw">names</span><span class="kw">=</span><span class="st">"TYMP"</span>))
<span class="co">## extract transcripts in the range</span>
<span class="no">trs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/geneModelFromTxdb.html">geneModelFromTxdb</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="no">org.Hs.eg.db</span>, <span class="kw">gr</span><span class="kw">=</span><span class="no">gr</span>)
<span class="co">## subset the features to show the interested transcripts only</span>
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu">GRangesList</span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">dat</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">dat</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">6</span>]]$<span class="no">dat</span>)
<span class="no">flen</span> <span class="kw">&lt;-</span> <span class="fu">elementNROWS</span>(<span class="no">features</span>)
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">features</span>)
<span class="co">## define the feature track layers</span>
<span class="no">features</span>$<span class="no">featureLayerID</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>:<span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">flen</span>[-<span class="fl">3</span>]), <span class="no">flen</span>[<span class="fl">3</span>]))
<span class="co">## define the feature labels</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">features</span>) <span class="kw">&lt;-</span> <span class="no">features</span>$<span class="no">symbol</span>
<span class="co">## define the feature colors</span>
<span class="no">features</span>$<span class="no">fill</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightblue"</span>, <span class="st">"mistyrose"</span>, <span class="st">"mistyrose"</span>), <span class="no">flen</span>)
<span class="co">## define the feature heights</span>
<span class="no">features</span>$<span class="no">height</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">features</span>$<span class="no">feature</span><span class="kw">==</span><span class="st">"CDS"</span>, <span class="fl">.04</span>, <span class="fl">.02</span>)
<span class="co">## import methylation data from a bed file</span>
<span class="no">methy</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"methy.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"trackViewer"</span>), <span class="st">"BED"</span>)
<span class="co">## subset the data to simplify information</span>
<span class="no">methy</span> <span class="kw">&lt;-</span> <span class="no">methy</span>[<span class="no">methy</span>$<span class="no">score</span> <span class="kw">&gt;</span> <span class="fl">20</span>]
<span class="co">## for pie plot, there are must be at least two numeric columns</span>
<span class="no">methy</span>$<span class="no">score2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">methy</span>$<span class="no">score</span>) - <span class="no">methy</span>$<span class="no">score</span>
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"methylated"</span>, <span class="st">"unmethylated"</span>),
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"lightblue"</span>),
                <span class="kw">color</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"black"</span>))
<span class="co">## lollipop plot, pie layout</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">methy</span>, <span class="no">features</span>,
          <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"pie"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/plotMethyBed-1.png" alt="lollipop plot, pie layout" width="864"><p class="caption">
lollipop plot, pie layout
</p>
</div>
</div>
<div id="plot-lollipop-plot-for-multiple-patients-in-pie-stack-layout" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-lollipop-plot-for-multiple-patients-in-pie-stack-layout" class="anchor"></a>Plot lollipop plot for multiple patients in “pie.stack” layout</h2>
<p>The percentage of methylation rates are shown by pie graph in different layers for different patients.</p>
<div class="sourceCode" id="cb100"><html><body><pre class="r"><span class="co">## simulate multiple patients</span>
<span class="no">rand.id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>), <span class="fl">3</span>*<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">rand.id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">rand.id</span>)
<span class="no">methy.mul.patient</span> <span class="kw">&lt;-</span> <span class="no">methy</span>[<span class="no">rand.id</span>]
<span class="co">## pie.stack require metadata "stack.factor", and the metadata can not be </span>
<span class="co">## stack.factor.order or stack.factor.first</span>
<span class="no">len.max</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">rand.id</span>))
<span class="no">stack.factors</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"patient"</span>,
                        <span class="fu"><a href="https://rdrr.io/r/base/formatc.html">formatC</a></span>(<span class="fl">1</span>:<span class="no">len.max</span>,
                                <span class="kw">width</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html">nchar</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">len.max</span>)),
                                <span class="kw">flag</span><span class="kw">=</span><span class="st">"0"</span>))
<span class="no">methy.mul.patient</span>$<span class="no">stack.factor</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">rand.id</span>), <span class="no">sample</span>, <span class="kw">x</span><span class="kw">=</span><span class="no">stack.factors</span>))
<span class="no">methy.mul.patient</span>$<span class="no">score</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy.mul.patient</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="co">## for a pie plot, two or more numeric meta-columns are required.</span>
<span class="no">methy.mul.patient</span>$<span class="no">score2</span> <span class="kw">&lt;-</span> <span class="fl">100</span> - <span class="no">methy.mul.patient</span>$<span class="no">score</span>
<span class="co">## set different color set for different patient</span>
<span class="no">patient.color.set</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(<span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">stack.factors</span>)),
                                                 <span class="st">"#FFFFFFFF"</span>),
                                           <span class="kw">stringsAsFactors</span><span class="kw">=</span><span class="fl">FALSE</span>))
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">patient.color.set</span>) <span class="kw">&lt;-</span> <span class="no">stack.factors</span>
<span class="no">methy.mul.patient</span>$<span class="no">color</span> <span class="kw">&lt;-</span>
    <span class="no">patient.color.set</span>[<span class="no">methy.mul.patient</span>$<span class="no">stack.factor</span>]
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="no">stack.factors</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"gray80"</span>,
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">patient.color.set</span>, <span class="no">`[`</span>, <span class="fl">1</span>))
<span class="co">## lollipop plot</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">methy.mul.patient</span>,
          <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>,
          <span class="kw">type</span><span class="kw">=</span><span class="st">"pie.stack"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/plotMethypieStack-1.png" alt="lollipop plot, pie.stack layout" width="864"><p class="caption">
lollipop plot, pie.stack layout
</p>
</div>
</div>
<div id="plot-lollipop-plot-in-caterpillar-layout-to-compare-two-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-lollipop-plot-in-caterpillar-layout-to-compare-two-samples" class="anchor"></a>Plot lollipop plot in caterpillar layout to compare two samples</h2>
<p>The caterpillar layout can be used to side by side comparing two samples or to display dense data.</p>
<div class="sourceCode" id="cb101"><html><body><pre class="r"><span class="co">## use SNPsideID to set the layer of event</span>
<span class="no">sample.gr</span>$<span class="no">SNPsideID</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"top"</span>, <span class="st">"bottom"</span>),
                               <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">sample.gr</span>),
                               <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">sample.gr</span>, <span class="no">features.gr</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/plotCaterpillar-1.png" alt="lollipop plot, caterpillar layout" width="864"><p class="caption">
lollipop plot, caterpillar layout
</p>
</div>
</div>
<div id="dandelion-plot-hundreds-snps" class="section level2">
<h2 class="hasAnchor">
<a href="#dandelion-plot-hundreds-snps" class="anchor"></a>Dandelion plot hundreds SNPs</h2>
<p>Sometimes, there are as many as hundreds of SNPs or methylation status involved in one gene. Dandelion plot can be used to depict such dense SNPs or methylation. Please note that the height of the dandelion indicates the density of the events.</p>
<div class="sourceCode" id="cb102"><html><body><pre class="r"><span class="no">methy</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"methy.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"trackViewer"</span>), <span class="st">"BED"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>)</pre></body></html></div>
<pre><code>## [1] 38</code></pre>
<div class="sourceCode" id="cb104"><html><body><pre class="r"><span class="co">## set the color of dandelion leaves.</span>
<span class="no">methy</span>$<span class="no">color</span> <span class="kw">&lt;-</span> <span class="fl">3</span>
<span class="no">methy</span>$<span class="no">border</span> <span class="kw">&lt;-</span> <span class="st">"gray"</span>
<span class="co">## we suppose the total event are same (methy + unmethy)</span>
<span class="co">## we rescale the methylation score by max value of the score </span>
<span class="no">m</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">methy</span>$<span class="no">score</span>)
<span class="no">methy</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="no">methy</span>$<span class="no">score</span>/<span class="no">m</span>
<span class="co"># The area of the fan indicate the percentage of methylation or rate of mutation.</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/dandelion.plot.html">dandelion.plot</a></span>(<span class="no">methy</span>, <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"fan"</span>)</pre></body></html></div>
<div class="figure">
<img src="ChIPpeakAnno_files/figure-html/dandelionPlot-1.png" alt="dandelion plot" width="576"><p class="caption">
dandelion plot
</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kai Hu, Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
