<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Integrated ChIP-seq Data Analysis Workshop • integratedchipseqanalysisworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Integrated ChIP-seq Data Analysis Workshop">
<meta property="og:description" content="integratedchipseqanalysisworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">integratedchipseqanalysisworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/IntegratedChIPseqAnalysisWorkshop.html">Integrated ChIP-seq Data Analysis Workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hukai916/IntegratedChIPseqWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Integrated ChIP-seq Data Analysis Workshop</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/hukai916/IntegratedChIPseqWorkshop/blob/master/vignettes/IntegratedChIPseqAnalysisWorkshop.Rmd"><code>vignettes/IntegratedChIPseqAnalysisWorkshop.Rmd</code></a></small>
      <div class="hidden name"><code>IntegratedChIPseqAnalysisWorkshop.Rmd</code></div>

    </div>

    
    
<p>The Docker container (<a href="https://hub.docker.com/repository/docker/hukai916/integratedchipseqanalysis_workshop" class="uri">https://hub.docker.com/repository/docker/hukai916/integratedchipseqanalysis_workshop</a>) with this workshop has pre-installed dependencies. If you are not using it, you must install the following packages.</p>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Load required libraries</h2>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span>({
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ChIPpeakAnno</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">trackViewer</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biomaRt</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">AnnotationHub</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">EnsDb.Hsapiens.v75</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome.Hsapiens.UCSC.hg19</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">UpSetR</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">seqinr</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">motifStack</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">WriteXLS</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rmarkdown</span>)
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">knitr</span>)})</pre></body></html></div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example1: annotate ChIP-seq peaks with ChIPpeakAnno</h2>
<p>Four steps:</p>
<ol style="list-style-type: decimal">
<li><p>Read in peak data with <code>toGRanges</code></p></li>
<li><p>Generate annotation data with <code>toGRanges</code></p></li>
<li><p>Annotate peaks with <code>annotatePeakInBatch</code></p></li>
<li><p>Add additional info with <code>addGeneIDs</code></p></li>
</ol>
<div id="my-section" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Step1: read in peak data</h3>
<p>We first need to convert peak files from bed/broadPeak/narrowPeak etc. format to <code>GRanges</code> object with <code>toGRanges</code> function.</p>
<p>For the demo, we will use the example data stored in ChIPpeakAnno package.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co">## locate example data:</span>
<span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"Tead4.broadPeak"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)

<span class="co">## read in peak file:</span>
<span class="no">peaks</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">path</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"broadPeak"</span>)

<span class="co">## inspect the top 2 lines of the peak file in GRanges format:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">peaks</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 4 metadata columns:
##             seqnames        ranges strand |     score signalValue    pValue
##                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;   &lt;numeric&gt; &lt;numeric&gt;
##   peak12338     chr2 175473-176697      * |       206      668.42        -1
##   peak12339     chr2 246412-246950      * |        31      100.23        -1
##                qValue
##             &lt;numeric&gt;
##   peak12338        -1
##   peak12339        -1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="my-section" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Step2: prepare annotation data</h3>
<p>Depending on your research goals, you can choose to use either ENSEMBL or UCSC annotations.</p>
<p>The more complex ENSEMBL annotation is preferred when the goal is to discover and explain unknown biological mechanisms.</p>
<p>The less complex UCSC annotation can be used to generate more robust and reproducible results.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)

<span class="no">annoDataTxDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)

<span class="co">## inspect the first two lines of the annotation file:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoDataTxDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 0 metadata columns:
##      seqnames            ranges strand
##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##    1    chr19 58858172-58874214      -
##   10     chr8 18248755-18258723      +
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome</code></pre>
</div>
<div id="my-section" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Step3: annotate peaks</h3>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co">## annotate by the nearest TSS based on UCSC hg19 annotations</span>
<span class="no">annoTxDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span>(<span class="no">peaks</span>, <span class="kw">AnnotationData</span><span class="kw">=</span><span class="no">annoDataTxDb</span>)

<span class="co">## inspect the first two lines of the annotation:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoTxDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 13 metadata columns:
##                   seqnames        ranges strand |     score signalValue
##                      &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;   &lt;numeric&gt;
##   peak12338.26751     chr2 175473-176697      * |       206      668.42
##   peak12339.26751     chr2 246412-246950      * |        31      100.23
##                      pValue    qValue        peak     feature start_position
##                   &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;      &lt;integer&gt;
##   peak12338.26751        -1        -1   peak12338       26751         218136
##   peak12339.26751        -1        -1   peak12339       26751         218136
##                   end_position feature_strand insideFeature distancetoFeature
##                      &lt;integer&gt;    &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
##   peak12338.26751       264810              -    downstream             89337
##   peak12339.26751       264810              -        inside             18398
##                   shortestDistance fromOverlappingOrNearest
##                          &lt;integer&gt;              &lt;character&gt;
##   peak12338.26751            41439          NearestLocation
##   peak12339.26751            17860          NearestLocation
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="my-section" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Step4: add additional annotation</h3>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>)

<span class="co">## add in gene symbol</span>
<span class="no">annoTxDb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/addGeneIDs.html">addGeneIDs</a></span>(<span class="no">annoTxDb</span>, <span class="kw">orgAnn</span><span class="kw">=</span><span class="st">"org.Hs.eg.db"</span>,
                        <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>,
                        <span class="kw">IDs2Add</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"symbol"</span>))
<span class="co">## inspect the first two lines:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">annoTxDb</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 14 metadata columns:
##                   seqnames        ranges strand |     score signalValue
##                      &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;   &lt;numeric&gt;
##   peak12338.26751     chr2 175473-176697      * |       206      668.42
##   peak12339.26751     chr2 246412-246950      * |        31      100.23
##                      pValue    qValue        peak     feature start_position
##                   &lt;numeric&gt; &lt;numeric&gt; &lt;character&gt; &lt;character&gt;      &lt;integer&gt;
##   peak12338.26751        -1        -1   peak12338       26751         218136
##   peak12339.26751        -1        -1   peak12339       26751         218136
##                   end_position feature_strand insideFeature distancetoFeature
##                      &lt;integer&gt;    &lt;character&gt;   &lt;character&gt;         &lt;numeric&gt;
##   peak12338.26751       264810              -    downstream             89337
##   peak12339.26751       264810              -        inside             18398
##                   shortestDistance fromOverlappingOrNearest      symbol
##                          &lt;integer&gt;              &lt;character&gt; &lt;character&gt;
##   peak12338.26751            41439          NearestLocation      SH3YL1
##   peak12339.26751            17860          NearestLocation      SH3YL1
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Short summary:</p>
<p>Be careful that peak annotation results may differ depending on the source of your annotation file.</p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example2: find overlaps for replicates</h2>
<p>The output <strong>overlappingPeaks</strong> object of the module <code>findOverlapsOfPeaks</code> consists of the followings:</p>
<ol style="list-style-type: decimal">
<li><p>venn_cnt</p></li>
<li><p>peaklist: a list of overlapping or unique peaks</p></li>
<li><p>overlappingPeaks: a list of data frame that consists of the annotations of all overlapping peaks</p></li>
</ol>
<p>Below, we would like to find the overlaps between two peak files: one from BED, the other from GFF.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ChIPpeakAnno</span>)

<span class="co">## convert BED file to GRanges:</span>
<span class="no">bed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"MACS_output.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="no">gr1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">bed</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"BED"</span>, <span class="kw">header</span><span class="kw">=</span><span class="fl">FALSE</span>)

<span class="co">## convert GFF file to GRanges:</span>
<span class="no">gff</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"GFF_peaks.gff"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="no">gr2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/toGRanges.html">toGRanges</a></span>(<span class="no">gff</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"GFF"</span>, <span class="kw">header</span><span class="kw">=</span><span class="fl">FALSE</span>, <span class="kw">skip</span><span class="kw">=</span><span class="fl">3</span>)

<span class="co">## must keep the class of gr2$score the same as gr1$score, i.e., numeric:</span>
<span class="no">gr2</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="no">gr2</span>$<span class="no">score</span>)

<span class="co">## find overlaps:</span>
<span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/findOverlapsOfPeaks.html">findOverlapsOfPeaks</a></span>(<span class="no">gr1</span>, <span class="no">gr2</span>, <span class="kw">connectedPeaks</span> <span class="kw">=</span> <span class="st">"keepAll"</span>)

<span class="co">## add metadata (mean of score) to the overlapping peaks</span>
<span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/addMetadata.html">addMetadata</a></span>(<span class="no">ol</span>, <span class="kw">colNames</span><span class="kw">=</span><span class="st">"score"</span>, <span class="kw">FUN</span><span class="kw">=</span><span class="no">mean</span>)

<span class="co">## inspect the first two lines of overlapping peaks:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">ol</span>$<span class="no">peaklist</span><span class="kw">[[</span><span class="st">"gr1///gr2"</span>]], <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames        ranges strand |                           peakNames
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |                     &lt;CharacterList&gt;
##   [1]     chr1 713791-715578      * | gr1__MACS_peak_13,gr2__001,gr2__002
##   [2]     chr1 724851-727191      * |          gr2__003,gr1__MACS_peak_14
##           score
##       &lt;numeric&gt;
##   [1]   850.203
##   [2]    29.170
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co">## make a Venn diagram:</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#009E73"</span>, <span class="st">"#F0E442"</span>),
                     <span class="kw">col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>),
                 <span class="kw">cat.col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>))</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/findOverlapsOfPeaks-1.png" alt="Venn diagram of overlaps for replicates" width="288"><p class="caption">
Venn diagram of overlaps for replicates
</p>
</div>
<pre><code>## $p.value
##      gr1 gr2 pval
## [1,]   1   1    0
## 
## $vennCounts
##      gr1 gr2 Counts count.gr1 count.gr2
## [1,]   0   0      0         0         0
## [2,]   0   1     61         0        61
## [3,]   1   0     62        62         0
## [4,]   1   1    166       168       169
## attr(,"class")
## [1] "VennCounts"</code></pre>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example3: visualize peak site distribution</h2>
<p>The <code>assignChromosomeRegion</code> function summarizes the distribution of peaks over various features (exon, intron, enhancer, promoter and UTR).</p>
<p>Results can be displayed in either peak-centric or nucleotide-centric view.</p>
<p>Note that a single peak might span multiple type of features.</p>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Binding site relative to features</h4>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)

<span class="co">## get overlapping peaks:</span>
<span class="no">overlaps</span> <span class="kw">&lt;-</span> <span class="no">ol</span>$<span class="no">peaklist</span><span class="kw">[[</span><span class="st">"gr1///gr2"</span>]]

<span class="co">## find peak site distribution:</span>
<span class="no">aCR</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/assignChromosomeRegion.html">assignChromosomeRegion</a></span>(<span class="no">overlaps</span>, <span class="kw">nucleotideLevel</span><span class="kw">=</span><span class="fl">FALSE</span>,
                              <span class="kw">precedence</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Promoters"</span>,
                                           <span class="st">"immediateDownstream"</span>,
                                           <span class="st">"fiveUTRs"</span>,
                                           <span class="st">"threeUTRs"</span>,
                                           <span class="st">"Exons"</span>,
                                           <span class="st">"Introns"</span>),
                              <span class="kw">TxDb</span><span class="kw">=</span><span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>)

<span class="co">## use barplot to visualize:</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span>(<span class="no">aCR</span>$<span class="no">percentage</span>, <span class="kw">las</span><span class="kw">=</span><span class="fl">3</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/assignChromosomeRegion-1.png" alt="Peak distribution over different genomic features." width="576"><p class="caption">
Peak distribution over different genomic features.
</p>
</div>
</div>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Peak distance distribution to nearest feature</h4>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/binOverFeature.html">binOverFeature</a></span>(<span class="no">overlaps</span>, <span class="kw">annotationData</span><span class="kw">=</span><span class="no">annoDataTxDb</span>,
               <span class="kw">radius</span><span class="kw">=</span><span class="fl">5000</span>, <span class="kw">nbins</span><span class="kw">=</span><span class="fl">20</span>, <span class="kw">FUN</span><span class="kw">=</span><span class="no">length</span>, <span class="kw">errFun</span><span class="kw">=</span><span class="fl">0</span>,
               <span class="kw">ylab</span><span class="kw">=</span><span class="st">"count"</span>,
               <span class="kw">main</span><span class="kw">=</span><span class="st">"Distribution of aggregated peak numbers around TSS"</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/binOverFeature-1.png" alt="Distribution of peaks around TSS." width="576"><p class="caption">
Distribution of peaks around TSS.
</p>
</div>
</div>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Annotate peaks to promoter regions (-500bp ~ +2000bp)</h4>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>)

<span class="co">## annotate to promoter regions:</span>
<span class="no">overlaps.anno</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/annotatePeakInBatch.html">annotatePeakInBatch</a></span>(<span class="no">overlaps</span>,
                                     <span class="kw">AnnotationData</span><span class="kw">=</span><span class="no">annoDataTxDb</span>,
                                     <span class="kw">output</span><span class="kw">=</span><span class="st">"nearestBiDirectionalPromoters"</span>,
                                     <span class="kw">bindingRegion</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(-<span class="fl">2000</span>, <span class="fl">500</span>))

<span class="co">## add in gene symbol:</span>
<span class="no">overlaps.anno</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/addGeneIDs.html">addGeneIDs</a></span>(<span class="no">overlaps.anno</span>,
                            <span class="st">"org.Hs.eg.db"</span>,
                            <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>,
                            <span class="kw">IDs2Add</span> <span class="kw">=</span> <span class="st">"symbol"</span>)

<span class="co">## visualized the first two lines:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">overlaps.anno</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRanges object with 2 ranges and 10 metadata columns:
##      seqnames        ranges strand |                           peakNames
##         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |                     &lt;CharacterList&gt;
##   X1     chr1 713791-715578      * | gr1__MACS_peak_13,gr2__001,gr2__002
##   X4     chr1 856361-856999      * |          gr1__MACS_peak_17,gr2__005
##          score        peak     feature feature.ranges feature.strand  distance
##      &lt;numeric&gt; &lt;character&gt; &lt;character&gt;      &lt;IRanges&gt;          &lt;Rle&gt; &lt;integer&gt;
##   X1   850.203          X1   100288069  700245-714068              -         0
##   X4    54.690          X4   100130417  852953-854817              -      1543
##      insideFeature distanceToSite       symbol
##        &lt;character&gt;      &lt;integer&gt;  &lt;character&gt;
##   X1  overlapStart              0 LOC100288069
##   X4      upstream           1543    LINC02593
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example4: find peaks with bi-directional promoters</h2>
<p>Bi-directional promoters refer to genomic regions that are located between TSS of two adjacent genes that are transcribed on opposite directions. Bi-directional promoters often regulate both genes.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="co">## find bi-directional promoters:</span>
<span class="no">bdp</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/peaksNearBDP.html">peaksNearBDP</a></span>(<span class="no">overlaps</span>, <span class="no">annoDataTxDb</span>, <span class="kw">maxgap</span><span class="kw">=</span><span class="fl">5000</span>)

<span class="co">## some statistics:</span>
<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">bdp</span>$<span class="no">percentPeaksWithBDP</span>,
  <span class="no">bdp</span>$<span class="no">n.peaks</span>,
  <span class="no">bdp</span>$<span class="no">n.peaksWithBDP</span>)</pre></body></html></div>
<pre><code>## [1]   0.05421687 166.00000000   9.00000000</code></pre>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co">## show first two lines:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">bdp</span>$<span class="no">peaksWithBDP</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## GRangesList object of length 2:
## $`4`
## GRanges object with 2 ranges and 10 metadata columns:
##      seqnames        ranges strand |                  peakNames     score
##         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;CharacterList&gt; &lt;numeric&gt;
##   X4     chr1 856361-856999      * | gr1__MACS_peak_17,gr2__005     54.69
##   X4     chr1 856361-856999      * | gr1__MACS_peak_17,gr2__005     54.69
##        bdp_idx        peak     feature feature.ranges feature.strand  distance
##      &lt;integer&gt; &lt;character&gt; &lt;character&gt;      &lt;IRanges&gt;          &lt;Rle&gt; &lt;integer&gt;
##   X4         4          X4   100130417  852953-854817              -      1543
##   X4         4          X4      148398  860530-879961              +      3530
##      insideFeature distanceToSite
##        &lt;character&gt;      &lt;integer&gt;
##   X4      upstream           1543
##   X4      upstream           3530
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
## 
## $`5`
## GRanges object with 2 ranges and 10 metadata columns:
##      seqnames        ranges strand |                  peakNames     score
##         &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;CharacterList&gt; &lt;numeric&gt;
##   X5     chr1 859315-860144      * | gr2__006,gr1__MACS_peak_18    81.485
##   X5     chr1 859315-860144      * | gr2__006,gr1__MACS_peak_18    81.485
##        bdp_idx        peak     feature feature.ranges feature.strand  distance
##      &lt;integer&gt; &lt;character&gt; &lt;character&gt;      &lt;IRanges&gt;          &lt;Rle&gt; &lt;integer&gt;
##   X5         5          X5   100130417  852953-854817              -      4497
##   X5         5          X5      148398  860530-879961              +       385
##      insideFeature distanceToSite
##        &lt;character&gt;      &lt;integer&gt;
##   X5      upstream           4497
##   X5      upstream            385
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example5: output a summary of peak consensus</h2>
<p>Many methods are available:</p>
<ol style="list-style-type: decimal">
<li><p>Use <code>getAllPeakSequence</code> module to output fastq file and sesarch the motif by other tools like MEME and HOMER.</p></li>
<li><p>Compare pre-defined patterns against target consensus patterns with <code>summarizePatternInPeaks</code> module.</p></li>
<li><p>Calculate the z-scores for all combinations of oligonucleotide at a given length with the <code>oligoSmmary</code> module.</p></li>
</ol>
<p>Below demonstrates method 3.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">seqinr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BSgenome.Hsapiens.UCSC.hg19</span>)

<span class="co">## obtain peak sequences:</span>
<span class="no">seq</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/getAllPeakSequence.html">getAllPeakSequence</a></span>(<span class="no">overlaps</span>,
                          <span class="kw">upstream</span><span class="kw">=</span><span class="fl">20</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">20</span>,
                          <span class="kw">genome</span><span class="kw">=</span><span class="no">Hsapiens</span>)

<span class="co">## output the fasta file for the 3nd program:</span>
<span class="co">#write2FASTA(seq, "test.fa")</span>

<span class="co">## summarize short oligos:</span>
<span class="no">os</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/oligoSummary.html">oligoSummary</a></span>(<span class="no">seq</span>, <span class="kw">oligoLength</span><span class="kw">=</span><span class="fl">6</span>, <span class="kw">MarkovOrder</span><span class="kw">=</span><span class="fl">3</span>, <span class="kw">quickMotif</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="co">## plot z-scores:</span>
<span class="no">zscore</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">os</span>$<span class="no">zscore</span>)
<span class="no">h</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="no">zscore</span>, <span class="kw">breaks</span><span class="kw">=</span><span class="fl">100</span>, <span class="kw">main</span><span class="kw">=</span><span class="st">"Histogram of Z-score"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(<span class="no">zscore</span>[<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">zscore</span>)], <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">h</span>$<span class="no">counts</span>)/<span class="fl">10</span>,
     <span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">zscore</span>[<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">zscore</span>)]), <span class="kw">srt</span><span class="kw">=</span><span class="fl">90</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/consensus-1.png" alt="Histogram of Z-score of 6-mer" width="576"><p class="caption">
Histogram of Z-score of 6-mer
</p>
</div>
<p>Generate motif using simulated data.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">motifStack</span>)

<span class="co">## generate the motifs:</span>
<span class="no">pfms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html">mapply</a></span>(<span class="kw">function</span>(<span class="no">.ele</span>, <span class="no">id</span>)
               <span class="fu">new</span>(<span class="st">"pfm"</span>, <span class="kw">mat</span><span class="kw">=</span><span class="no">.ele</span>, <span class="kw">name</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"SAMPLE motif"</span>, <span class="no">id</span>)),
               <span class="no">os</span>$<span class="no">motifs</span>, <span class="fl">1</span>:<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">os</span>$<span class="no">motifs</span>))
<span class="fu"><a href="https://rdrr.io/pkg/motifStack/man/motifStack.html">motifStack</a></span>(<span class="no">pfms</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/motif-1.png" alt="Motif of simulation data" width="576"><p class="caption">
Motif of simulation data
</p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example6: obtain enriched GO and pathways</h2>
<p>Please note that by default, _feature_id|<em>type</em> is set to “ensembl_gene_id”, is you are using <strong>TxDb</strong> for your annotation, please set to “entrez_id”.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reactome.db</span>)

<span class="co">## obtain enriched GO term:</span>
<span class="no">over</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/getEnrichedGO.html">getEnrichedGO</a></span>(<span class="no">overlaps.anno</span>, <span class="kw">orgAnn</span><span class="kw">=</span><span class="st">"org.Hs.eg.db"</span>,
                      <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>,
                      <span class="kw">maxP</span><span class="kw">=</span><span class="fl">.05</span>, <span class="kw">minGOterm</span><span class="kw">=</span><span class="fl">10</span>,
                      <span class="kw">multiAdjMethod</span><span class="kw">=</span><span class="st">"BH"</span>, <span class="kw">condense</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="co">## show first two lines:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">over</span><span class="kw">[[</span><span class="st">"mf"</span>]][, -<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>, <span class="fl">10</span>)], <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>## [1] go.id              go.term            Ontology           pvalue            
## [5] count.InDataset    count.InGenome     totaltermInDataset totaltermInGenome 
## [9] EntrezID          
## &lt;0 rows&gt; (or 0-length row.names)</code></pre>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="co">## obtain enriched pathways:</span>
<span class="no">pathway</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/getEnrichedPATH.html">getEnrichedPATH</a></span>(<span class="no">overlaps.anno</span>, <span class="st">"org.Hs.eg.db"</span>, <span class="st">"reactome.db"</span>,
                        <span class="kw">feature_id_type</span><span class="kw">=</span><span class="st">"entrez_id"</span>, <span class="kw">maxP</span><span class="kw">=</span><span class="fl">.05</span>)

<span class="co">## show the first two lines:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">pathway</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">2</span>)</pre></body></html></div>
<pre><code>##        path.id EntrezID count.InDataset count.InGenome     pvalue
## 1 R-HSA-196849    55229               2            123 0.03029926
## 2 R-HSA-196849    65220               2            123 0.03029926
##   totaltermInDataset totaltermInGenome PATH
## 1                244            111075   NA
## 2                244            111075   NA</code></pre>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example7: determine if there is significant peak overlap among multiple sets</h2>
<p>Given multiple peak lists, to see if the binding sites are correlated and what is the common binding pattern. Two methods: hypergeometric test and permutation test.</p>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Method 1: hypergeometric test</h4>
<p>This method requires the number of all potential binding sites to be known. You need to set the <em>totalTest</em> (total potential peak number) in the <code>makeVennDiagram</code> module. The value should be larger than the max number of peaks in the peak list.</p>
<p>In the example below, we assume a 3% coding region plus promoters. Since the sample data is only a subset of chr2, we estimate that the total binding sites is 1/24 of possible binding region in the genome.</p>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="co">## read in data: TAF, Tead4 and YY1</span>
<span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">path</span>, <span class="st">"broadPeak"</span>)
<span class="no">data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="no">files</span>), <span class="no">toGRanges</span>, <span class="kw">format</span><span class="kw">=</span><span class="st">"broadPeak"</span>)
(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">data</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".broadPeak"</span>, <span class="st">""</span>, <span class="no">files</span>))</pre></body></html></div>
<pre><code>## [1] "TAF"   "Tead4" "YY1"</code></pre>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="co">## find overlapped peaks:</span>
<span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/findOverlapsOfPeaks.html">findOverlapsOfPeaks</a></span>(<span class="no">data</span>, <span class="kw">connectedPeaks</span><span class="kw">=</span><span class="st">"keepAll"</span>)

<span class="co">## avergage of peak width:</span>
<span class="no">averagePeakWidth</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="fu">width</span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu">GRangesList</span>(<span class="no">ol</span>$<span class="no">peaklist</span>))))

<span class="co">## estimate of total binding site;</span>
<span class="no">tot</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span>(<span class="fl">3.3e+9</span> * <span class="fl">.03</span> / <span class="no">averagePeakWidth</span> / <span class="fl">24</span>)

<span class="co">## make Venn diagram:</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">totalTest</span><span class="kw">=</span><span class="no">tot</span>, <span class="kw">connectedPeaks</span><span class="kw">=</span><span class="st">"keepAll"</span>,
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#CC79A7"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#F0E442"</span>),
                <span class="kw">col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>),
                <span class="kw">cat.col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>))</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/vennDiagram-1.png" alt="Venn diagram of overlaps." width="384"><p class="caption">
Venn diagram of overlaps.
</p>
</div>
<pre><code>## $p.value
##      TAF Tead4 YY1          pval
## [1,]   0     1   1  1.000000e+00
## [2,]   1     0   1 2.904297e-258
## [3,]   1     1   0  8.970986e-04
## 
## $vennCounts
##      TAF Tead4 YY1 Counts count.TAF count.Tead4 count.YY1
## [1,]   0     0   0    849         0           0         0
## [2,]   0     0   1    621         0           0       621
## [3,]   0     1   0   2097         0        2097         0
## [4,]   0     1   1    309         0         310       319
## [5,]   1     0   0     59        59           0         0
## [6,]   1     0   1    166       172           0       170
## [7,]   1     1   0      8         8           8         0
## [8,]   1     1   1    476       545         537       521
## attr(,"class")
## [1] "VennCounts"</code></pre>
<p>Keep first list consistent.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="co">## set connectedPeaks to keepFirstListConsistent will show consistent total number of peaks for the first peak list.</span>
<span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/makeVennDiagram.html">makeVennDiagram</a></span>(<span class="no">ol</span>, <span class="kw">totalTest</span><span class="kw">=</span><span class="no">tot</span>, <span class="kw">connectedPeaks</span><span class="kw">=</span><span class="st">"keepFirstListConsistent"</span>,
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#CC79A7"</span>, <span class="st">"#56B4E9"</span>, <span class="st">"#F0E442"</span>),
                <span class="kw">col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>),
                <span class="kw">cat.col</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#D55E00"</span>, <span class="st">"#0072B2"</span>, <span class="st">"#E69F00"</span>))</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/vennDiagramForFirstConsistent-1.png" alt="Venn diagram of overlaps for first TF." width="384"><p class="caption">
Venn diagram of overlaps for first TF.
</p>
</div>
<pre><code>## $p.value
##      TAF Tead4 YY1          pval
## [1,]   0     1   1  1.000000e+00
## [2,]   1     0   1 2.904297e-258
## [3,]   1     1   0  8.970986e-04
## 
## $vennCounts
##      TAF Tead4 YY1 Counts count.TAF count.Tead4 count.YY1
## [1,]   0     0   0    849         0           0         0
## [2,]   0     0   1    621         0           0       621
## [3,]   0     1   0   2097         0        2097         0
## [4,]   0     1   1    309         0         310       319
## [5,]   1     0   0     59        59           0         0
## [6,]   1     0   1    166       172           0       170
## [7,]   1     1   0      8         8           8         0
## [8,]   1     1   1    476       545         537       521
## attr(,"class")
## [1] "VennCounts"</code></pre>
</div>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Method 2: Permutation test</h4>
<p>With <code>peakPermTest</code>, users don’t need to estimate total potential binding sites.</p>
<p>Before permutation test, random peaks need to be generated using the distribution discovered from the input peak for a given feature type. Alternatively, a peak pool can be prepared with <code>preparePool</code> module. Below is an exmple using the transcription factor binding site clusters (V3) (see ?<code>wgEncodeTfbsV3</code>) downloaded from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeRegTfbsClusteredV3.bed.gz">ENCODE</a> with the HOT spots (?<code>HOT.spots</code>) removed.</p>
<p>HOT spots are regions with probability to be bound by many TFs. We suggest to remove them before the test. Users can also choose to remove blacklist <a href="https://sites.google.com/site/anshulkundaje/projects/blacklists">ENCODE blacklist</a> regions. Note that some blacklists need to be converted to the correct genome assembly with liftover utility.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="co">## read in data:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">HOT.spots</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">wgEncodeTfbsV3</span>)
<span class="no">hotGR</span> <span class="kw">&lt;-</span> <span class="fu">reduce</span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">HOT.spots</span>))

<span class="co">## remove HOT spots:</span>
<span class="no">removeOl</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">.ele</span>){
    <span class="no">ol</span> <span class="kw">&lt;-</span> <span class="fu">findOverlaps</span>(<span class="no">.ele</span>, <span class="no">hotGR</span>)
    <span class="kw">if</span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">ol</span>)<span class="kw">&gt;</span><span class="fl">0</span>) <span class="no">.ele</span> <span class="kw">&lt;-</span> <span class="no">.ele</span>[-<span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu">queryHits</span>(<span class="no">ol</span>))]
    <span class="no">.ele</span>
}
<span class="no">TAF</span> <span class="kw">&lt;-</span> <span class="fu">removeOl</span>(<span class="no">data</span><span class="kw">[[</span><span class="st">"TAF"</span>]])
<span class="no">TEAD4</span> <span class="kw">&lt;-</span> <span class="fu">removeOl</span>(<span class="no">data</span><span class="kw">[[</span><span class="st">"Tead4"</span>]])
<span class="no">YY1</span> <span class="kw">&lt;-</span> <span class="fu">removeOl</span>(<span class="no">data</span><span class="kw">[[</span><span class="st">"YY1"</span>]])

<span class="co">## subset to save demo time:</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">wgEncodeTfbsV3.subset</span> <span class="kw">&lt;-</span> <span class="no">wgEncodeTfbsV3</span>[<span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">wgEncodeTfbsV3</span>), <span class="fl">2000</span>)]

<span class="co">## YY1 vs TEAD4:</span>
<span class="no">pool</span> <span class="kw">&lt;-</span> <span class="fu">new</span>(<span class="st">"permPool"</span>, <span class="kw">grs</span><span class="kw">=</span><span class="fu">GRangesList</span>(<span class="no">wgEncodeTfbsV3.subset</span>), <span class="kw">N</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">YY1</span>))
<span class="no">pt1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/peakPermTest.html">peakPermTest</a></span>(<span class="no">YY1</span>, <span class="no">TEAD4</span>, <span class="kw">pool</span><span class="kw">=</span><span class="no">pool</span>, <span class="kw">seed</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">force.parallel</span><span class="kw">=</span><span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">pt1</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/peakPermTest-1.png" alt="permutation test for YY1 and TEAD4" width="576"><p class="caption">
permutation test for YY1 and TEAD4
</p>
</div>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="co">## YY1 vs TAF:</span>
<span class="no">pt2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/peakPermTest.html">peakPermTest</a></span>(<span class="no">YY1</span>, <span class="no">TAF</span>, <span class="kw">pool</span><span class="kw">=</span><span class="no">pool</span>, <span class="kw">seed</span><span class="kw">=</span><span class="fl">1</span>, <span class="kw">force.parallel</span><span class="kw">=</span><span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">pt2</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/peakPermTest2-1.png" alt="permutation test for YY1 and TAF" width="576"><p class="caption">
permutation test for YY1 and TAF
</p>
</div>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example8: metagene analysis for a given feature/peak range</h2>
<p>You can easily visualize and compare the binding patterns of raw signals of multiple ChIP-seq experiments using function <code>featureAlignedHeatmap</code> and <code>featureAlignedDistribution</code>.</p>
<p>Heatmap of aligned features.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)

<span class="co">## recenter features:</span>
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="no">ol</span>$<span class="no">peaklist</span><span class="kw">[[</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">ol</span>$<span class="no">peaklist</span>)]]
<span class="no">feature.recentered</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html">reCenterPeaks</a></span>(<span class="no">features</span>, <span class="kw">width</span><span class="kw">=</span><span class="fl">4000</span>)

<span class="co">## read in data:</span>
<span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"ChIPpeakAnno"</span>)
<span class="no">files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="no">path</span>, <span class="st">"bigWig"</span>)
<span class="kw">if</span>(<span class="no">.Platform</span>$<span class="no">OS.type</span> <span class="kw">!=</span> <span class="st">"windows"</span>){
    <span class="no">cvglists</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="no">files</span>), <span class="no">import</span>,
                       <span class="kw">format</span><span class="kw">=</span><span class="st">"BigWig"</span>,
                       <span class="kw">which</span><span class="kw">=</span><span class="no">feature.recentered</span>,
                       <span class="kw">as</span><span class="kw">=</span><span class="st">"RleList"</span>)
} <span class="kw">else</span> {
    <span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="no">path</span>, <span class="st">"cvglist.rds"</span>))
}

<span class="co">## extract signal:</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">cvglists</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">".bigWig"</span>, <span class="st">""</span>, <span class="no">files</span>)
<span class="no">feature.center</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/reCenterPeaks.html">reCenterPeaks</a></span>(<span class="no">features</span>, <span class="kw">width</span><span class="kw">=</span><span class="fl">1</span>)
<span class="no">sig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedSignal.html">featureAlignedSignal</a></span>(<span class="no">cvglists</span>, <span class="no">feature.center</span>,
                            <span class="kw">upstream</span><span class="kw">=</span><span class="fl">2000</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">2000</span>)

<span class="co">## output heatmap</span>
<span class="no">keep</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="no">sig</span><span class="kw">[[</span><span class="fl">2</span>]]) <span class="kw">&gt;</span> <span class="fl">0</span>
<span class="no">sig</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">sig</span>, <span class="kw">function</span>(<span class="no">.ele</span>) <span class="no">.ele</span>[<span class="no">keep</span>, ], <span class="kw">simplify</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="no">feature.center</span> <span class="kw">&lt;-</span> <span class="no">feature.center</span>[<span class="no">keep</span>]
<span class="no">heatmap</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedHeatmap.html">featureAlignedHeatmap</a></span>(<span class="no">sig</span>, <span class="no">feature.center</span>,
                                 <span class="kw">upstream</span><span class="kw">=</span><span class="fl">2000</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">2000</span>,
                                 <span class="kw">upper.extreme</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">3</span>,<span class="fl">.5</span>,<span class="fl">4</span>))</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/heatmap-1.png" alt="Heatmap of aligned features sorted by signal of TAF" width="384"><p class="caption">
Heatmap of aligned features sorted by signal of TAF
</p>
</div>
<p>Distriution of aligned features.</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/ChIPpeakAnno/man/featureAlignedDistribution.html">featureAlignedDistribution</a></span>(<span class="no">sig</span>, <span class="no">feature.center</span>,
                           <span class="kw">upstream</span><span class="kw">=</span><span class="fl">2000</span>, <span class="kw">downstream</span><span class="kw">=</span><span class="fl">2000</span>,
                           <span class="kw">type</span><span class="kw">=</span><span class="st">"l"</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/distribution-1.png" alt="Distribution of aligned features" width="480"><p class="caption">
Distribution of aligned features
</p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example9: creating lollipop plot with trackViewer</h2>
<p>It is routine to show the distribution of mutation or genetic variations by lollipop-style (or needle-style) plot in a genome browser, along with a variety of genomic annotations, such as gene or exon level models, CpG island, and so on.</p>
<p>For SNP status or methlyation data, lollipop plot is often used. Many tools can provide lollipop plot such as cBioPortal Tools::MutationMapper<span class="citation"><sup>1</sup></span>, EMBL-EBI::Pfam<span class="citation"><sup>2</sup></span>, and BioJS::muts-needle-plot<span class="citation"><sup>3</sup></span>, BiQ Analyzer<span class="citation"><sup>4</sup></span>, and Methylation plotter<span class="citation"><sup>5</sup></span>. The cBioPortal Tools::MutationMapper is a well-known and easy to use online genome browser that can generate high quality figures with mutations by inputing tab-delimited mutation data. In R/Bioconductor, with the Rgraphics system, there are many flexible ways to display methylation data such as <em>MethVisual</em><span class="citation"><sup>6</sup></span>, <em>REDseq</em><span class="citation"><sup>7</sup></span>, and <em>GenVisR</em><span class="citation"><sup>8</sup></span>.</p>
<p>Below is a summary of current tools that are available.</p>
<table class="table">
<colgroup>
<col width="44%">
<col width="19%">
<col width="7%">
<col width="28%">
</colgroup>
<thead><tr class="header">
<th align="left">software</th>
<th align="left">inputs</th>
<th align="left">online</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MutationMapper<span class="citation"><sup>1</sup></span>
</td>
<td align="left">tab-delimited text</td>
<td align="left">Yes</td>
<td align="left">interprets mutations with different heights along protein annotations in automatic color theme</td>
</tr>
<tr class="even">
<td align="left">Pfam<span class="citation"><sup>2</sup></span>
</td>
<td align="left">JSON</td>
<td align="left">Yes</td>
<td align="left">could combine different line and head colors with different drawing styles</td>
</tr>
<tr class="odd">
<td align="left">muts-needle-plot<span class="citation"><sup>3</sup></span>
</td>
<td align="left">JSON</td>
<td align="left">No</td>
<td align="left">plots data point with different colors, heights, and size along annotations, and highlighted selcted coordinates</td>
</tr>
<tr class="even">
<td align="left">BiQ Analyzer<span class="citation"><sup>4</sup></span>
</td>
<td align="left">BiQ methylation file</td>
<td align="left">Yes</td>
<td align="left">interprets methylation status in black &amp; white</td>
</tr>
<tr class="odd">
<td align="left">Methylation plotter<span class="citation"><sup>5</sup></span>
</td>
<td align="left">tab-delimited text</td>
<td align="left">Yes</td>
<td align="left">stacked multiple methylation status for multiple samples</td>
</tr>
<tr class="even">
<td align="left">
<em>MethVisual</em><span class="citation"><sup>6</sup></span>
</td>
<td align="left">R list</td>
<td align="left">No</td>
<td align="left">visualize the methylation status of CpGs according to their genomic position</td>
</tr>
<tr class="odd">
<td align="left">
<em>REDseq</em><span class="citation"><sup>7</sup></span>
</td>
<td align="left">R list</td>
<td align="left">No</td>
<td align="left">plot frequencies of methylations and SNPs along a chromosome</td>
</tr>
<tr class="even">
<td align="left">
<em>GenVisR</em><span class="citation"><sup>8</sup></span>
</td>
<td align="left">R dataframe</td>
<td align="left">No</td>
<td align="left">plot most accurate graphic representation of the ensembl annotation version based on biomart service</td>
</tr>
</tbody>
</table>
<p>Though the tools above meet the basic even advanced visualization requirement, they do have limitations. For example, it will be hard to display the data if there are multiple mutations in the same posistion. Moreover, generating high quality pictures in bunch for publication is also a pain.</p>
<p>To fill this gap, we developed <em>trackViewer</em>, a R/Bioconductor package serving as an enhanced light-weight genome viewer for visualizing various types of high-throughput sequencing data, side by side or stacked with multiple annotation tracks.</p>
<p>Besides the regular read coverage tracks supported by existing genome browsers/viewers, <em>trackViewer</em> can also generate lollipop plot to depict the methylation and SNP/mutation status, together with coverage data and annotation tracks to facilitate integrated analysis of multi-omics data. In addition, figures generated by <em>trackViewer</em> are interactive, i.e., the feel-and-look such as the layout, the color scheme and the order of tracks can be easily customized by the users. Furthermore, <em>trackViewer</em> can be easily integrated into standard analysis pipeline for various high-throughput sequencing dataset such as ChIP-seq, RNA-seq, methylation-seq or DNA-seq. The images produced by <em>trackViewer</em> are highly customizable including labels, symbols, colors and size. Here, we illustrate its utilities and capabilities in deriving biological insights from multi-omics dataset from GEO/ENCODE.</p>
<p>There are 3 steps for generating lollipop plot:</p>
<ol style="list-style-type: decimal">
<li><p>Prepare the methylation/variant/mutation data.</p></li>
<li><p>Prepare the gene/protein model.</p></li>
<li><p>Plot lollipop plot.</p></li>
</ol>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Step1: repare the methylation/variant/mutation data.</h4>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">trackViewer</span>)

<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>)

<span class="co"># Here we use SNP sample data</span>
<span class="no">SNP</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">10</span>, <span class="fl">100</span>, <span class="fl">105</span>, <span class="fl">108</span>, <span class="fl">400</span>, <span class="fl">410</span>, <span class="fl">420</span>, <span class="fl">600</span>, <span class="fl">700</span>, <span class="fl">805</span>, <span class="fl">840</span>, <span class="fl">1400</span>, <span class="fl">1402</span>)
<span class="co"># use GenomicRanges::GRanges function to create a GRanges object.</span>
<span class="co"># for real data, users can import vcf data via VariantAnnotation::readVcf function.</span>
<span class="no">sample.gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr1"</span>, <span class="fu">IRanges</span>(<span class="no">SNP</span>, <span class="kw">width</span><span class="kw">=</span><span class="fl">1</span>,
                     <span class="co">## the name of GRanges object will be used as label</span>
                     <span class="kw">names</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"snp"</span>, <span class="no">SNP</span>)),
                     <span class="co">## score value will be used to for the height of lollipop</span>
                     <span class="kw">score</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                     <span class="co">## set the color for lollipop node.</span>
                     <span class="kw">color</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">6</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                     <span class="co">## set the lollipop stem color</span>
                     <span class="kw">border</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"gray80"</span>, <span class="st">"gray30"</span>),
                     <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>))</pre></body></html></div>
</div>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Step2: prepare the gene/protein model.</h4>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="no">features.gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr1"</span>, <span class="fu">IRanges</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">501</span>, <span class="fl">1001</span>),
                                       <span class="kw">width</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">120</span>, <span class="fl">400</span>, <span class="fl">405</span>),
                                       <span class="kw">names</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"exon"</span>, <span class="fl">1</span>:<span class="fl">3</span>)),
                                                     <span class="kw">fill</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#FF8833"</span>, <span class="st">"#51C6E6"</span>, <span class="st">"#DFA32D"</span>), <span class="co">## color for exon</span>
                                                     <span class="kw">height</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.02</span>, <span class="fl">0.05</span>, <span class="fl">0.08</span>) <span class="co">## height for exon</span>
                        )</pre></body></html></div>
</div>
<div id="my-section" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Step3: plot lollipop plot</h4>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">sample.gr</span>, <span class="no">features.gr</span>)</pre></body></html></div>
<p><img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/quickStartStep3-1.png" width="576"></p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example10: plot from VCF file with trackViewer</h2>
<p>VCF is a text file format that contains metadata and mutation information about genomic positions, original genotypes and optional genotypes.</p>
<p>The <em>trackViewer</em> package can show single nucleotide polymorphisms (SNPs) from VCF file in lollipop plot. Figure @ref(fig:plotVCFdata) shows an example lollipop plot of real SNPs. Sample SNPs are a subset of 1000 variants and 50 samples from chromosome 22 taken from 1000 Genomes in VCF in the <em>VariantAnnotation</em> package. Different colors depict the new SNP events in the circles. The number of circles indicates the number of SNP events.</p>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">VariantAnnotation</span>) <span class="co">## load package for reading vcf file</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>) <span class="co">## load package for gene model</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>) <span class="co">## load package for gene name</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)

<span class="no">fl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"chr22.vcf.gz"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"VariantAnnotation"</span>)
<span class="co">## set the track range</span>
<span class="no">gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"22"</span>, <span class="fu">IRanges</span>(<span class="fl">50968014</span>, <span class="fl">50970514</span>, <span class="kw">names</span><span class="kw">=</span><span class="st">"TYMP"</span>))
<span class="co">## read in vcf file</span>
<span class="no">tab</span> <span class="kw">&lt;-</span> <span class="fu">TabixFile</span>(<span class="no">fl</span>)
<span class="no">vcf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(<span class="no">fl</span>, <span class="st">"hg19"</span>, <span class="kw">param</span><span class="kw">=</span><span class="no">gr</span>)

<span class="co">## get GRanges from VCF object</span>
<span class="no">mutation.frequency</span> <span class="kw">&lt;-</span> <span class="fu">rowRanges</span>(<span class="no">vcf</span>)
<span class="co">## keep the metadata</span>
<span class="fu">mcols</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="fu">mcols</span>(<span class="no">mutation.frequency</span>),
          <span class="kw pkg">VariantAnnotation</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html">info</a></span>(<span class="no">vcf</span>))

<span class="co">## set colors</span>
<span class="no">mutation.frequency</span>$<span class="no">border</span> <span class="kw">&lt;-</span> <span class="st">"gray30"</span>
<span class="no">mutation.frequency</span>$<span class="no">color</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"^rs"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">mutation.frequency</span>)),
           <span class="st">"lightcyan"</span>, <span class="st">"lavender"</span>)

<span class="co">## plot Global Allele Frequency based on AC/AN</span>
<span class="no">mutation.frequency</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="no">mutation.frequency</span>$<span class="no">AF</span>*<span class="fl">100</span>)
<span class="co">## change the SNPs label rotation angle</span>
<span class="no">mutation.frequency</span>$<span class="no">label.parameter.rot</span> <span class="kw">&lt;-</span> <span class="fl">45</span>
<span class="co">## keep sequence level style same</span>
<span class="fu">seqlevelsStyle</span>(<span class="no">gr</span>) <span class="kw">&lt;-</span> <span class="fu">seqlevelsStyle</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span> <span class="st">"UCSC"</span>
<span class="fu">seqlevels</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span> <span class="st">"chr22"</span>

<span class="co">## extract transcripts in the range</span>
<span class="no">trs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/geneModelFromTxdb.html">geneModelFromTxdb</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="no">org.Hs.eg.db</span>, <span class="kw">gr</span><span class="kw">=</span><span class="no">gr</span>)

<span class="co">## subset the features to show the interested transcripts only</span>
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">dat</span>), <span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">dat</span>))
<span class="co">## define the feature labels</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">features</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">name</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">name</span>)
<span class="co">## define the feature colors</span>
<span class="no">features</span>$<span class="no">fill</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightblue"</span>, <span class="st">"mistyrose"</span>)
<span class="co">## define the feature heights</span>
<span class="no">features</span>$<span class="no">height</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">.02</span>, <span class="fl">.04</span>)
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"known"</span>, <span class="st">"unkown"</span>),
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightcyan"</span>, <span class="st">"lavender"</span>),
                <span class="kw">color</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gray80"</span>, <span class="st">"gray80"</span>))
<span class="co">## lollipop plot</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">mutation.frequency</span>,
          <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"circle"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/plotVCFdata-1.png" alt="lollipop plot for VCF data" width="864"><p class="caption">
lollipop plot for VCF data
</p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example11: plot methylation data from BED file with trackViewer</h2>
<p>Any type of data that can be imported into a <code>GRanges</code> object can be viewed by <em>trackViewer</em> package.</p>
<p>Sample methylations are random data generated for illustration and are saved in BED format file. The <em>rtracklayer</em> package is used to import the methylation data into a <code>GRanges</code>. The transcripts are extracted from <code>TxDb</code> object and are assigned gene symbol with <code>org</code> database. We also demonstrate that multiple transcripts can be shown in different colors and tracks.</p>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>) <span class="co">## load package for gene model</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>) <span class="co">## load package for gene name</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)

<span class="co">## set the track range</span>
<span class="no">gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr22"</span>, <span class="fu">IRanges</span>(<span class="fl">50968014</span>, <span class="fl">50970514</span>, <span class="kw">names</span><span class="kw">=</span><span class="st">"TYMP"</span>))
<span class="co">## extract transcripts in the range</span>
<span class="no">trs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/geneModelFromTxdb.html">geneModelFromTxdb</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="no">org.Hs.eg.db</span>, <span class="kw">gr</span><span class="kw">=</span><span class="no">gr</span>)
<span class="co">## subset the features to show the interested transcripts only</span>
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu">GRangesList</span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">dat</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">dat</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">6</span>]]$<span class="no">dat</span>)
<span class="no">flen</span> <span class="kw">&lt;-</span> <span class="fu">elementNROWS</span>(<span class="no">features</span>)
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">features</span>)

<span class="co">## define the feature track layers</span>
<span class="no">features</span>$<span class="no">featureLayerID</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>:<span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">flen</span>[-<span class="fl">3</span>]), <span class="no">flen</span>[<span class="fl">3</span>]))
<span class="co">## define the feature labels</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">features</span>) <span class="kw">&lt;-</span> <span class="no">features</span>$<span class="no">symbol</span>
<span class="co">## define the feature colors</span>
<span class="no">features</span>$<span class="no">fill</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightblue"</span>, <span class="st">"mistyrose"</span>, <span class="st">"mistyrose"</span>), <span class="no">flen</span>)
<span class="co">## define the feature heights</span>
<span class="no">features</span>$<span class="no">height</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">features</span>$<span class="no">feature</span><span class="kw">==</span><span class="st">"CDS"</span>, <span class="fl">.04</span>, <span class="fl">.02</span>)
<span class="co">## import methylation data from a bed file</span>
<span class="no">methy</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"methy.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"trackViewer"</span>), <span class="st">"BED"</span>)
<span class="co">## subset the data to simplify information</span>
<span class="no">methy</span> <span class="kw">&lt;-</span> <span class="no">methy</span>[<span class="no">methy</span>$<span class="no">score</span> <span class="kw">&gt;</span> <span class="fl">20</span>]
<span class="co">## for pie plot, there are must be at least two numeric columns</span>
<span class="no">methy</span>$<span class="no">score2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">methy</span>$<span class="no">score</span>) - <span class="no">methy</span>$<span class="no">score</span>
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"methylated"</span>, <span class="st">"unmethylated"</span>),
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"lightblue"</span>),
                <span class="kw">color</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"black"</span>))

<span class="co">## lollipop plot, pie layout</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">methy</span>, <span class="no">features</span>,
          <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"pie"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/plotMethyBed-1.png" alt="lollipop plot, pie layout" width="864"><p class="caption">
lollipop plot, pie layout
</p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example12: plot lollipop plot for multiple patients in “pie.stack” layout with trackViewer</h2>
<p>The percentage of methylation rates are shown by pie graph in different layers for different patients.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="co">## simulate multiple patients</span>
<span class="no">rand.id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>), <span class="fl">3</span>*<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">rand.id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">rand.id</span>)
<span class="no">methy.mul.patient</span> <span class="kw">&lt;-</span> <span class="no">methy</span>[<span class="no">rand.id</span>]

<span class="co">## pie.stack require metadata "stack.factor", and the metadata can not be stack.factor.order or stack.factor.first</span>
<span class="no">len.max</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">rand.id</span>))
<span class="no">stack.factors</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"patient"</span>,
                        <span class="fu"><a href="https://rdrr.io/r/base/formatc.html">formatC</a></span>(<span class="fl">1</span>:<span class="no">len.max</span>,
                                <span class="kw">width</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html">nchar</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">len.max</span>)),
                                <span class="kw">flag</span><span class="kw">=</span><span class="st">"0"</span>))
<span class="no">methy.mul.patient</span>$<span class="no">stack.factor</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">rand.id</span>), <span class="no">sample</span>, <span class="kw">x</span><span class="kw">=</span><span class="no">stack.factors</span>))
<span class="no">methy.mul.patient</span>$<span class="no">score</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy.mul.patient</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="co">## for a pie plot, two or more numeric meta-columns are required.</span>
<span class="no">methy.mul.patient</span>$<span class="no">score2</span> <span class="kw">&lt;-</span> <span class="fl">100</span> - <span class="no">methy.mul.patient</span>$<span class="no">score</span>
<span class="co">## set different color set for different patient</span>
<span class="no">patient.color.set</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(<span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">stack.factors</span>)),
                                                 <span class="st">"#FFFFFFFF"</span>),
                                           <span class="kw">stringsAsFactors</span><span class="kw">=</span><span class="fl">FALSE</span>))
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">patient.color.set</span>) <span class="kw">&lt;-</span> <span class="no">stack.factors</span>
<span class="no">methy.mul.patient</span>$<span class="no">color</span> <span class="kw">&lt;-</span>
    <span class="no">patient.color.set</span>[<span class="no">methy.mul.patient</span>$<span class="no">stack.factor</span>]

<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="no">stack.factors</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"gray80"</span>,
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">patient.color.set</span>, <span class="no">`[`</span>, <span class="fl">1</span>))

<span class="co">## lollipop plot</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">methy.mul.patient</span>,
          <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>,
          <span class="kw">type</span><span class="kw">=</span><span class="st">"pie.stack"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/plotMethypieStack-1.png" alt="lollipop plot, pie.stack layout" width="864"><p class="caption">
lollipop plot, pie.stack layout
</p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example13: plot lollipop plot in caterpillar layout to compare two samples with trackViewer</h2>
<p>The caterpillar layout can be used to compare two samples or to display dense data side by side.</p>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="co">## use SNPsideID to set the layer of event</span>
<span class="no">sample.gr</span>$<span class="no">SNPsideID</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"top"</span>, <span class="st">"bottom"</span>),
                               <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">sample.gr</span>),
                               <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">sample.gr</span>, <span class="no">features.gr</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/plotCaterpillar-1.png" alt="lollipop plot, caterpillar layout" width="864"><p class="caption">
lollipop plot, caterpillar layout
</p>
</div>
</div>
<div id="my-section" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#my-section" class="anchor"></a>Example14: dandelion plot hundreds SNPs with trackViewer</h2>
<p>Sometimes, there are as many as hundreds of SNPs or methylation status involved in one gene. Dandelion plot can be used to depict such dense SNPs or methylation. Please note that the height of the dandelion indicates the density of the events.</p>
<div class="sourceCode" id="cb45"><html><body><pre class="r"><span class="no">methy</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"methy.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"trackViewer"</span>), <span class="st">"BED"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>)</pre></body></html></div>
<pre><code>## [1] 38</code></pre>
<div class="sourceCode" id="cb47"><html><body><pre class="r"><span class="co">## set the color of dandelion leaves.</span>
<span class="no">methy</span>$<span class="no">color</span> <span class="kw">&lt;-</span> <span class="fl">3</span>
<span class="no">methy</span>$<span class="no">border</span> <span class="kw">&lt;-</span> <span class="st">"gray"</span>

<span class="co">## we suppose the total event are same (methy + unmethy)</span>
<span class="co">## we rescale the methylation score by max value of the score</span>
<span class="no">m</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">methy</span>$<span class="no">score</span>)
<span class="no">methy</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="no">methy</span>$<span class="no">score</span>/<span class="no">m</span>

<span class="co"># The area of the fan indicate the percentage of methylation or rate of mutation.</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/dandelion.plot.html">dandelion.plot</a></span>(<span class="no">methy</span>, <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"fan"</span>)</pre></body></html></div>
<div class="figure">
<img src="IntegratedChIPseqAnalysisWorkshop_files/figure-html/dandelionPlot-1.png" alt="dandelion plot" width="576"><p class="caption">
dandelion plot
</p>
</div>
<div id="refs" class="references">
<div id="ref-gao2013integrative">
<p>1. Gao, J. <em>et al.</em> Integrative analysis of complex cancer genomics and clinical profiles using the cBioPortal. <em>Science signaling</em> <strong>6,</strong> pl1 (2013).</p>
</div>
<div id="ref-finn2006pfam">
<p>2. Finn, R. D. <em>et al.</em> Pfam: Clans, web tools and services. <em>Nucleic acids research</em> <strong>34,</strong> D247–D251 (2006).</p>
</div>
<div id="ref-michael_p_schroeder_2015_14561">
<p>3. Schroeder, M. P. Muts-needle-plot: Mutations needle plot v0.8.0. (2015). doi:<a href="https://doi.org/10.5281/zenodo.14561">10.5281/zenodo.14561</a></p>
</div>
<div id="ref-bock2005biq">
<p>4. Bock, C. <em>et al.</em> BiQ analyzer: Visualization and quality control for dna methylation data from bisulfite sequencing. <em>Bioinformatics</em> <strong>21,</strong> 4067–4068 (2005).</p>
</div>
<div id="ref-mallona2014methylation">
<p>5. Mallona, I., Díez-Villanueva, A. &amp; Peinado, M. A. Methylation plotter: A web tool for dynamic visualization of dna methylation data. <em>Source code for biology and medicine</em> <strong>9,</strong> 1 (2014).</p>
</div>
<div id="ref-zackay2010methvisual">
<p>6. Zackay, A. &amp; Steinhoff, C. MethVisual-visualization and exploratory statistical analysis of dna methylation profiles from bisulfite sequencing. <em>BMC research notes</em> <strong>3,</strong> 337 (2010).</p>
</div>
<div id="ref-zhu4redseq">
<p>7. Zhu, L. &amp; others. REDseq: A bioconductor package for analyzing high throughput sequencing data from restriction enzyme digestion.</p>
</div>
<div id="ref-skidmore2016genvisr">
<p>8. Skidmore, Z. L. <em>et al.</em> GenVisR: Genomic visualizations in r. <em>Bioinformatics</em> btw325 (2016).</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kai Hu, Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
