<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>trackViewer_workshop2020 • integratedchipseqanalysisworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="trackViewer_workshop2020">
<meta property="og:description" content="integratedchipseqanalysisworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">integratedchipseqanalysisworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/trackViewer.html">trackViewer_workshop2020</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hukai916/IntegratedChIPseqWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="trackViewer_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>trackViewer_workshop2020</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/hukai916/IntegratedChIPseqWorkshop/blob/master/vignettes/trackViewer.Rmd"><code>vignettes/trackViewer.Rmd</code></a></small>
      <div class="hidden name"><code>trackViewer.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>It is a routine way to show the distribution of mutation for genetic variations by lollipop-style (or needle-style) plot in a genome browser, along with a variety of genomic annotations, such as gene level or exon level models, CpG island, and so on. To show the SNP status or methylation data, a special plot, called lollipop plot or needle plot, is used to show the distribution along annotations. Many of currently available tools offline or online provide lollipop-like plot to show the mutation data such as cBioPortal Tools::MutationMapper<span class="citation"><sup>1</sup></span>, EMBL-EBI::Pfam<span class="citation"><sup>2</sup></span>, and BioJS::muts-needle-plot<span class="citation"><sup>3</sup></span>, BiQ Analyzer<span class="citation"><sup>4</sup></span>, and Methylation plotter<span class="citation"><sup>5</sup></span>. The cBioPortal Tools::MutationMapper is a well-known and easy to use on-line genome browser that can output high quality of figures with mutations by input tab-delimited mutation data. In R/Bioconductor, there are options to use the flexibility of the Rgraphics system to display Methylation data such as <em>MethVisual</em><span class="citation"><sup>6</sup></span>, <em>REDseq</em><span class="citation"><sup>7</sup></span>, and <em>GenVisR</em><span class="citation"><sup>8</sup></span>.</p>
<p><strong>Table 1: Tools availble for lollipop plot. </strong></p>
<table class="table">
<colgroup>
<col width="44%">
<col width="19%">
<col width="7%">
<col width="28%">
</colgroup>
<thead><tr class="header">
<th align="left">software</th>
<th align="left">inputs</th>
<th align="left">online</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MutationMapper<span class="citation"><sup>1</sup></span>
</td>
<td align="left">tab-delimited text</td>
<td align="left">Yes</td>
<td align="left">interprets mutations with different heights along protein annotations in automatic color theme</td>
</tr>
<tr class="even">
<td align="left">Pfam<span class="citation"><sup>2</sup></span>
</td>
<td align="left">JSON</td>
<td align="left">Yes</td>
<td align="left">could combine different line and head colors with different drawing styles</td>
</tr>
<tr class="odd">
<td align="left">muts-needle-plot<span class="citation"><sup>3</sup></span>
</td>
<td align="left">JSON</td>
<td align="left">No</td>
<td align="left">plots data point with different colors, heights, and size along annotations, and highlighted selcted coordinates</td>
</tr>
<tr class="even">
<td align="left">BiQ Analyzer<span class="citation"><sup>4</sup></span>
</td>
<td align="left">BiQ methylation file</td>
<td align="left">Yes</td>
<td align="left">interprets methylation status in black &amp; white</td>
</tr>
<tr class="odd">
<td align="left">Methylation plotter<span class="citation"><sup>5</sup></span>
</td>
<td align="left">tab-delimited text</td>
<td align="left">Yes</td>
<td align="left">stacked multiple methylation status for multiple samples</td>
</tr>
<tr class="even">
<td align="left">
<em>MethVisual</em><span class="citation"><sup>6</sup></span>
</td>
<td align="left">R list</td>
<td align="left">No</td>
<td align="left">visualize the methylation status of CpGs according to their genomic position</td>
</tr>
<tr class="odd">
<td align="left">
<em>REDseq</em><span class="citation"><sup>7</sup></span>
</td>
<td align="left">R list</td>
<td align="left">No</td>
<td align="left">plot frequencies of methylations and SNPs along a chromosome</td>
</tr>
<tr class="even">
<td align="left">
<em>GenVisR</em><span class="citation"><sup>8</sup></span>
</td>
<td align="left">R dataframe</td>
<td align="left">No</td>
<td align="left">plot most accurate graphic representation of the ensembl annotation version based on biomart service</td>
</tr>
</tbody>
</table>
<p>All the tools available for the genomic data visualization mentioned above can meet the basic even complicated requirement. However, if there are multiple mutations in almost same position, it is very difficult to display the data using any available tools. What’s more there is a tendency that the figures it generated become more and more complex and busy. Thereby make it difficult in generating high quality pictures for publication in bunch. To fill this gap, we developed <em>trackViewer</em>, a R/Bioconductor package as an enhanced light-weight genome viewer for visualizing various types of high-throughput sequencing data, side by side or stacked with multiple annotation tracks. Besides the regular read coverage tracks supported by existing genome browsers/viewers, <em>trackViewer</em> can also be used to generate lollipop plot to depict the methylation and SNP/mutation status, together with coverage data and annotation tracks to facilitate integrated analysis of multi-omics data. In addition, figures generated by <em>trackViewer</em> are interactive, i.e., the feel-and-look such as the layout, the color scheme and the order of tracks can be easily customized by the users. Furthermore, <em>trackViewer</em> can be easily integrated into standard analysis pipeline for various high-throughput sequencing dataset such as ChIP-seq, RNA-seq, methylation-seq or DNA-seq. The images produced by <em>trackViewer</em> are highly customizable including labels, symbols, colors and size. Here, we illustrate its utilities and capabilities in deriving biological insights from multi-omics dataset from GEO/Encode.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><em>BiocManager</em> is used to install the released version of <em>trackViewer</em>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)
<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"trackViewer"</span>)</pre></body></html></div>
<p>To install the development version of <em>trackViewer</em>, please try</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">BiocManager</span>)
<span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"jianhong/trackViewer"</span>)</pre></body></html></div>
<p>To check the current version of <em>trackViewer</em>, please try</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/packageDescription.html">packageVersion</a></span>(<span class="st">"trackViewer"</span>)</pre></body></html></div>
<pre><code>## [1] '1.25.1'</code></pre>
</div>
<div id="quick-start-for-lollipop-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-start-for-lollipop-plot" class="anchor"></a>Quick start for lollipop plot</h2>
<div id="step-1--prepare-the-methylationvariantmutation-data" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1--prepare-the-methylationvariantmutation-data" class="anchor"></a>Step 1. Prepare the methylation/variant/mutation data</h3>
<p>The input data for lollipop plot by <em>trackViewer</em> is an object of <em>GenomicRanges</em>::<code>GRanges</code>.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co">## load library</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">trackViewer</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>) <span class="co">## set seed for random sampling to make sure it can be repeat.</span>
<span class="co">## Here we use SNP sample data</span>
<span class="no">SNP</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">10</span>, <span class="fl">100</span>, <span class="fl">105</span>, <span class="fl">108</span>, <span class="fl">400</span>, <span class="fl">410</span>, <span class="fl">420</span>, <span class="fl">600</span>, <span class="fl">700</span>, <span class="fl">805</span>, <span class="fl">840</span>, <span class="fl">1400</span>, <span class="fl">1402</span>)
<span class="co">## use GenomicRanges::GRanges function to create a GRanges object.</span>
<span class="co">## for real data, users can import vcf data via VariantAnnotation::readVcf function.</span>
<span class="no">sample.gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr1"</span>, <span class="fu">IRanges</span>(<span class="no">SNP</span>, <span class="kw">width</span><span class="kw">=</span><span class="fl">1</span>,
                                     <span class="co">## the name of GRanges object will be used as label</span>
                                     <span class="kw">names</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"snp"</span>, <span class="no">SNP</span>)),
                     <span class="co">## score value will be used to for the height of lollipop</span>
                     <span class="kw">score</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">5</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                     <span class="co">## set the color for lollipop node.</span>
                     <span class="kw">color</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">6</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                     <span class="co">## set the lollipop stem color</span>
                     <span class="kw">border</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"gray80"</span>, <span class="st">"gray30"</span>),
                                     <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">SNP</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
                     )</pre></body></html></div>
</div>
<div id="step-2--prepare-the-geneprotein-model" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2--prepare-the-geneprotein-model" class="anchor"></a>Step 2. Prepare the gene/protein model</h3>
<p>The gene/protein model for lollipop plot by <em>trackViewer</em> is also an object of <em>GenomicRanges</em>::<code>GRanges</code>.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">features.gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr1"</span>, <span class="fu">IRanges</span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>, <span class="fl">501</span>, <span class="fl">1001</span>),
                                       <span class="kw">width</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">120</span>, <span class="fl">400</span>, <span class="fl">405</span>),
                                       <span class="kw">names</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"exon"</span>, <span class="fl">1</span>:<span class="fl">3</span>)),
                       <span class="kw">fill</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"#FF8833"</span>, <span class="st">"#51C6E6"</span>, <span class="st">"#DFA32D"</span>), <span class="co">## color for exon</span>
                       <span class="kw">height</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.02</span>, <span class="fl">0.05</span>, <span class="fl">0.08</span>) <span class="co">## height for exon</span>
                    )</pre></body></html></div>
</div>
<div id="step-3--plot-lollipop-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3--plot-lollipop-plot" class="anchor"></a>Step 3. Plot lollipop plot</h3>
<p>The clustered events could be visualized one by one by jittered lollipop positions.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">sample.gr</span>, <span class="no">features.gr</span>)</pre></body></html></div>
<p><img src="trackViewer_files/figure-html/quickStartStep3-1.png" width="576"></p>
</div>
</div>
<div id="plot-data-from-variant-call-format-vcf-file" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-data-from-variant-call-format-vcf-file" class="anchor"></a>Plot data from Variant Call Format (VCF) file</h2>
<p>VCF is a text file format that contains metadata and mutation information about genomic positions, original genotypes and optional genotypes. The <em>trackViewer</em> package could show single nucleotide polymorphisms (SNPs) from VCF file in lollipop-style plot. Figure @ref(fig:plotVCFdata) shows an example lollipop plot of real SNPs. Sample SNPs are a subset of 1000 variants and 50 samples from chromosome 22 taken from 1000 Genomes in VCF in the <em>VariantAnnotation</em> package. Different colors depict the new SNP events in the circles. The number of circles indicates the number of SNP events.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">VariantAnnotation</span>) <span class="co">## load package for reading vcf file</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>) <span class="co">## load package for gene model</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>) <span class="co">## load package for gene name</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)
<span class="no">fl</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"chr22.vcf.gz"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"VariantAnnotation"</span>)
<span class="co">## set the track range</span>
<span class="no">gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"22"</span>, <span class="fu">IRanges</span>(<span class="fl">50968014</span>, <span class="fl">50970514</span>, <span class="kw">names</span><span class="kw">=</span><span class="st">"TYMP"</span>))
<span class="co">## read in vcf file</span>
<span class="no">tab</span> <span class="kw">&lt;-</span> <span class="fu">TabixFile</span>(<span class="no">fl</span>)
<span class="no">vcf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html">readVcf</a></span>(<span class="no">fl</span>, <span class="st">"hg19"</span>, <span class="kw">param</span><span class="kw">=</span><span class="no">gr</span>)
<span class="co">## get GRanges from VCF object </span>
<span class="no">mutation.frequency</span> <span class="kw">&lt;-</span> <span class="fu">rowRanges</span>(<span class="no">vcf</span>)
<span class="co">## keep the metadata</span>
<span class="fu">mcols</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="fu">mcols</span>(<span class="no">mutation.frequency</span>),
          <span class="kw pkg">VariantAnnotation</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html">info</a></span>(<span class="no">vcf</span>))
<span class="co">## set colors</span>
<span class="no">mutation.frequency</span>$<span class="no">border</span> <span class="kw">&lt;-</span> <span class="st">"gray30"</span>
<span class="no">mutation.frequency</span>$<span class="no">color</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"^rs"</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">mutation.frequency</span>)),
           <span class="st">"lightcyan"</span>, <span class="st">"lavender"</span>)
<span class="co">## plot Global Allele Frequency based on AC/AN</span>
<span class="no">mutation.frequency</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span>(<span class="no">mutation.frequency</span>$<span class="no">AF</span>*<span class="fl">100</span>)
<span class="co">## change the SNPs label rotation angle</span>
<span class="no">mutation.frequency</span>$<span class="no">label.parameter.rot</span> <span class="kw">&lt;-</span> <span class="fl">45</span>
<span class="co">## keep sequence level style same</span>
<span class="fu">seqlevelsStyle</span>(<span class="no">gr</span>) <span class="kw">&lt;-</span> <span class="fu">seqlevelsStyle</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span> <span class="st">"UCSC"</span>
<span class="fu">seqlevels</span>(<span class="no">mutation.frequency</span>) <span class="kw">&lt;-</span> <span class="st">"chr22"</span>
<span class="co">## extract transcripts in the range</span>
<span class="no">trs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/geneModelFromTxdb.html">geneModelFromTxdb</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="no">org.Hs.eg.db</span>, <span class="kw">gr</span><span class="kw">=</span><span class="no">gr</span>)
<span class="co">## subset the features to show the interested transcripts only</span>
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">dat</span>), <span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">dat</span>))
<span class="co">## define the feature labels</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">features</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">name</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">name</span>)
<span class="co">## define the feature colors</span>
<span class="no">features</span>$<span class="no">fill</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightblue"</span>, <span class="st">"mistyrose"</span>)
<span class="co">## define the feature heights</span>
<span class="no">features</span>$<span class="no">height</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">.02</span>, <span class="fl">.04</span>)
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"known"</span>, <span class="st">"unkown"</span>),
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightcyan"</span>, <span class="st">"lavender"</span>),
                <span class="kw">color</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gray80"</span>, <span class="st">"gray80"</span>))
<span class="co">## lollipop plot</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">mutation.frequency</span>,
          <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"circle"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="trackViewer_files/figure-html/plotVCFdata-1.png" alt="lollipop plot for VCF data" width="864"><p class="caption">
lollipop plot for VCF data
</p>
</div>
</div>
<div id="plot-methylation-data-from-bed-format-file" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-methylation-data-from-bed-format-file" class="anchor"></a>Plot methylation data from bed format file</h2>
<p>Any data can imported into a <code>GRanges</code> object can be viewed by <em>trackViewer</em> package. Sample methylations are random data generated for illustration and saved in bed format file. The <em>rtracklayer</em> package is used to import the methylation data into a <code>GRanges</code>. The transcripts are extracted from <code>TxDb</code> object and assigned gene symbol with <code>org</code> database. Here also show multiple transcripts could be shown in different colors and tracks.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>) <span class="co">## load package for gene model</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">org.Hs.eg.db</span>) <span class="co">## load package for gene name</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">rtracklayer</span>)
<span class="co">## set the track range</span>
<span class="no">gr</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="st">"chr22"</span>, <span class="fu">IRanges</span>(<span class="fl">50968014</span>, <span class="fl">50970514</span>, <span class="kw">names</span><span class="kw">=</span><span class="st">"TYMP"</span>))
<span class="co">## extract transcripts in the range</span>
<span class="no">trs</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/geneModelFromTxdb.html">geneModelFromTxdb</a></span>(<span class="no">TxDb.Hsapiens.UCSC.hg19.knownGene</span>,
                         <span class="no">org.Hs.eg.db</span>, <span class="kw">gr</span><span class="kw">=</span><span class="no">gr</span>)
<span class="co">## subset the features to show the interested transcripts only</span>
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu">GRangesList</span>(<span class="no">trs</span><span class="kw">[[</span><span class="fl">1</span>]]$<span class="no">dat</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">5</span>]]$<span class="no">dat</span>, <span class="no">trs</span><span class="kw">[[</span><span class="fl">6</span>]]$<span class="no">dat</span>)
<span class="no">flen</span> <span class="kw">&lt;-</span> <span class="fu">elementNROWS</span>(<span class="no">features</span>)
<span class="no">features</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">features</span>)
<span class="co">## define the feature track layers</span>
<span class="no">features</span>$<span class="no">featureLayerID</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">1</span>:<span class="fl">2</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(<span class="no">flen</span>[-<span class="fl">3</span>]), <span class="no">flen</span>[<span class="fl">3</span>]))
<span class="co">## define the feature labels</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">features</span>) <span class="kw">&lt;-</span> <span class="no">features</span>$<span class="no">symbol</span>
<span class="co">## define the feature colors</span>
<span class="no">features</span>$<span class="no">fill</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"lightblue"</span>, <span class="st">"mistyrose"</span>, <span class="st">"mistyrose"</span>), <span class="no">flen</span>)
<span class="co">## define the feature heights</span>
<span class="no">features</span>$<span class="no">height</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span>(<span class="no">features</span>$<span class="no">feature</span><span class="kw">==</span><span class="st">"CDS"</span>, <span class="fl">.04</span>, <span class="fl">.02</span>)
<span class="co">## import methylation data from a bed file</span>
<span class="no">methy</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"methy.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"trackViewer"</span>), <span class="st">"BED"</span>)
<span class="co">## subset the data to simplify information</span>
<span class="no">methy</span> <span class="kw">&lt;-</span> <span class="no">methy</span>[<span class="no">methy</span>$<span class="no">score</span> <span class="kw">&gt;</span> <span class="fl">20</span>]
<span class="co">## for pie plot, there are must be at least two numeric columns</span>
<span class="no">methy</span>$<span class="no">score2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">methy</span>$<span class="no">score</span>) - <span class="no">methy</span>$<span class="no">score</span>
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"methylated"</span>, <span class="st">"unmethylated"</span>),
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"white"</span>, <span class="st">"lightblue"</span>),
                <span class="kw">color</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"black"</span>, <span class="st">"black"</span>))
<span class="co">## lollipop plot, pie layout</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">methy</span>, <span class="no">features</span>,
          <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"pie"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="trackViewer_files/figure-html/plotMethyBed-1.png" alt="lollipop plot, pie layout" width="864"><p class="caption">
lollipop plot, pie layout
</p>
</div>
</div>
<div id="plot-lollipop-plot-for-multiple-patients-in-pie-stack-layout" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-lollipop-plot-for-multiple-patients-in-pie-stack-layout" class="anchor"></a>Plot lollipop plot for multiple patients in “pie.stack” layout</h2>
<p>The percentage of methylation rates are shown by pie graph in different layers for different patients.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co">## simulate multiple patients</span>
<span class="no">rand.id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>), <span class="fl">3</span>*<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="no">rand.id</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="no">rand.id</span>)
<span class="no">methy.mul.patient</span> <span class="kw">&lt;-</span> <span class="no">methy</span>[<span class="no">rand.id</span>]
<span class="co">## pie.stack require metadata "stack.factor", and the metadata can not be </span>
<span class="co">## stack.factor.order or stack.factor.first</span>
<span class="no">len.max</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">rand.id</span>))
<span class="no">stack.factors</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"patient"</span>,
                        <span class="fu"><a href="https://rdrr.io/r/base/formatc.html">formatC</a></span>(<span class="fl">1</span>:<span class="no">len.max</span>,
                                <span class="kw">width</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html">nchar</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">len.max</span>)),
                                <span class="kw">flag</span><span class="kw">=</span><span class="st">"0"</span>))
<span class="no">methy.mul.patient</span>$<span class="no">stack.factor</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">rand.id</span>), <span class="no">sample</span>, <span class="kw">x</span><span class="kw">=</span><span class="no">stack.factors</span>))
<span class="no">methy.mul.patient</span>$<span class="no">score</span> <span class="kw">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample.int</a></span>(<span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy.mul.patient</span>), <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="co">## for a pie plot, two or more numeric meta-columns are required.</span>
<span class="no">methy.mul.patient</span>$<span class="no">score2</span> <span class="kw">&lt;-</span> <span class="fl">100</span> - <span class="no">methy.mul.patient</span>$<span class="no">score</span>
<span class="co">## set different color set for different patient</span>
<span class="no">patient.color.set</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(<span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">rainbow</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">stack.factors</span>)),
                                                 <span class="st">"#FFFFFFFF"</span>),
                                           <span class="kw">stringsAsFactors</span><span class="kw">=</span><span class="fl">FALSE</span>))
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">patient.color.set</span>) <span class="kw">&lt;-</span> <span class="no">stack.factors</span>
<span class="no">methy.mul.patient</span>$<span class="no">color</span> <span class="kw">&lt;-</span>
    <span class="no">patient.color.set</span>[<span class="no">methy.mul.patient</span>$<span class="no">stack.factor</span>]
<span class="co">## set the legends</span>
<span class="no">legends</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">labels</span><span class="kw">=</span><span class="no">stack.factors</span>, <span class="kw">col</span><span class="kw">=</span><span class="st">"gray80"</span>,
                <span class="kw">fill</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">patient.color.set</span>, <span class="no">`[`</span>, <span class="fl">1</span>))
<span class="co">## lollipop plot</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">methy.mul.patient</span>,
          <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>,
          <span class="kw">type</span><span class="kw">=</span><span class="st">"pie.stack"</span>,
          <span class="kw">legend</span><span class="kw">=</span><span class="no">legends</span>)</pre></body></html></div>
<div class="figure">
<img src="trackViewer_files/figure-html/plotMethypieStack-1.png" alt="lollipop plot, pie.stack layout" width="864"><p class="caption">
lollipop plot, pie.stack layout
</p>
</div>
</div>
<div id="plot-lollipop-plot-in-caterpillar-layout-to-compare-two-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-lollipop-plot-in-caterpillar-layout-to-compare-two-samples" class="anchor"></a>Plot lollipop plot in caterpillar layout to compare two samples</h2>
<p>The caterpillar layout can be used to side by side comparing two samples or to display dense data.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co">## use SNPsideID to set the layer of event</span>
<span class="no">sample.gr</span>$<span class="no">SNPsideID</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"top"</span>, <span class="st">"bottom"</span>),
                               <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">sample.gr</span>),
                               <span class="kw">replace</span><span class="kw">=</span><span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/lolliplot.html">lolliplot</a></span>(<span class="no">sample.gr</span>, <span class="no">features.gr</span>)</pre></body></html></div>
<div class="figure">
<img src="trackViewer_files/figure-html/plotCaterpillar-1.png" alt="lollipop plot, caterpillar layout" width="864"><p class="caption">
lollipop plot, caterpillar layout
</p>
</div>
</div>
<div id="dandelion-plot-hundreds-snps" class="section level2">
<h2 class="hasAnchor">
<a href="#dandelion-plot-hundreds-snps" class="anchor"></a>Dandelion plot hundreds SNPs</h2>
<p>Sometimes, there are as many as hundreds of SNPs or methylation status involved in one gene. Dandelion plot can be used to depict such dense SNPs or methylation. Please note that the height of the dandelion indicates the density of the events.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">methy</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"methy.bed"</span>, <span class="kw">package</span><span class="kw">=</span><span class="st">"trackViewer"</span>), <span class="st">"BED"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">methy</span>)</pre></body></html></div>
<pre><code>## [1] 38</code></pre>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co">## set the color of dandelion leaves.</span>
<span class="no">methy</span>$<span class="no">color</span> <span class="kw">&lt;-</span> <span class="fl">3</span>
<span class="no">methy</span>$<span class="no">border</span> <span class="kw">&lt;-</span> <span class="st">"gray"</span>
<span class="co">## we suppose the total event are same (methy + unmethy)</span>
<span class="co">## we rescale the methylation score by max value of the score </span>
<span class="no">m</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">methy</span>$<span class="no">score</span>)
<span class="no">methy</span>$<span class="no">score</span> <span class="kw">&lt;-</span> <span class="no">methy</span>$<span class="no">score</span>/<span class="no">m</span>
<span class="co"># The area of the fan indicate the percentage of methylation or rate of mutation.</span>
<span class="fu"><a href="https://rdrr.io/pkg/trackViewer/man/dandelion.plot.html">dandelion.plot</a></span>(<span class="no">methy</span>, <span class="no">features</span>, <span class="kw">ranges</span><span class="kw">=</span><span class="no">gr</span>, <span class="kw">type</span><span class="kw">=</span><span class="st">"fan"</span>)</pre></body></html></div>
<div class="figure">
<img src="trackViewer_files/figure-html/dandelionPlot-1.png" alt="dandelion plot" width="576"><p class="caption">
dandelion plot
</p>
</div>
<div id="refs" class="references">
<div id="ref-gao2013integrative">
<p>1. Gao, J. <em>et al.</em> Integrative analysis of complex cancer genomics and clinical profiles using the cBioPortal. <em>Science signaling</em> <strong>6,</strong> pl1 (2013).</p>
</div>
<div id="ref-finn2006pfam">
<p>2. Finn, R. D. <em>et al.</em> Pfam: Clans, web tools and services. <em>Nucleic acids research</em> <strong>34,</strong> D247–D251 (2006).</p>
</div>
<div id="ref-michael_p_schroeder_2015_14561">
<p>3. Schroeder, M. P. Muts-needle-plot: Mutations needle plot v0.8.0. (2015). doi:<a href="https://doi.org/10.5281/zenodo.14561">10.5281/zenodo.14561</a></p>
</div>
<div id="ref-bock2005biq">
<p>4. Bock, C. <em>et al.</em> BiQ analyzer: Visualization and quality control for dna methylation data from bisulfite sequencing. <em>Bioinformatics</em> <strong>21,</strong> 4067–4068 (2005).</p>
</div>
<div id="ref-mallona2014methylation">
<p>5. Mallona, I., Dı́ez-Villanueva, A. &amp; Peinado, M. A. Methylation plotter: A web tool for dynamic visualization of dna methylation data. <em>Source code for biology and medicine</em> <strong>9,</strong> 1 (2014).</p>
</div>
<div id="ref-zackay2010methvisual">
<p>6. Zackay, A. &amp; Steinhoff, C. MethVisual-visualization and exploratory statistical analysis of dna methylation profiles from bisulfite sequencing. <em>BMC research notes</em> <strong>3,</strong> 337 (2010).</p>
</div>
<div id="ref-zhu4redseq">
<p>7. Zhu, L. &amp; others. REDseq: A bioconductor package for analyzing high throughput sequencing data from restriction enzyme digestion.</p>
</div>
<div id="ref-skidmore2016genvisr">
<p>8. Skidmore, Z. L. <em>et al.</em> GenVisR: Genomic visualizations in r. <em>Bioinformatics</em> btw325 (2016).</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kai Hu, Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
